--- /tmp/ffmpeg-6.0-src/fftools/ffmpeg.c	2026-04-06 17:05:14
+++ .fftools-7.1-work/kit60/ffmpeg.c	2026-04-06 18:08:33
@@ -1,5 +1,7 @@
 /*
  * Copyright (c) 2000-2003 Fabrice Bellard
+ * Copyright (c) 2018-2022 Taner Sener
+ * Copyright (c) 2023 ARTHENICA LTD
  *
  * This file is part of FFmpeg.
  *
@@ -23,6 +25,68 @@
  * multimedia converter based on the FFmpeg libraries
  */
 
+/*
+ * This file is the modified version of ffmpeg.c file living in ffmpeg source code under the fftools folder. We
+ * manually update it each time we depend on a new ffmpeg version. Below you can see the list of changes applied
+ * by us to develop mobile-ffmpeg and later ffmpeg-kit libraries.
+ *
+ * ffmpeg-kit changes by ARTHENICA LTD
+ *
+ * 07.2023
+ * --------------------------------------------------------
+ * - FFmpeg 6.0 changes migrated
+ * - vstats_file, received_sigterm and received_nb_signals updated as thread-local
+ * - forward_report method signature updated
+ * - time field in report_callback/forward_report/set_report_callback updated as double
+ *
+ * mobile-ffmpeg / ffmpeg-kit changes by Taner Sener
+ *
+ * 09.2022
+ * --------------------------------------------------------
+ * - added opt_common.h include
+ * - volatile dropped from thread local variables
+ * - setvbuf call dropped
+ * - flushing stderr dropped
+ * - muxing overhead printed in single line
+ *
+ * 08.2020
+ * --------------------------------------------------------
+ * - OptionDef defines combined
+ *
+ * 06.2020
+ * --------------------------------------------------------
+ * - ignoring signals implemented
+ * - cancel_operation() method signature updated with id
+ * - cancel by execution id implemented
+ * - volatile modifier added to critical variables
+ *
+ * 01.2020
+ * --------------------------------------------------------
+ * - ffprobe support (added ffmpeg_ prefix to methods and variables defined for both ffmpeg and ffprobe)
+ *
+ * 12.2019
+ * --------------------------------------------------------
+ * - concurrent execution support ("__thread" specifier added to variables used by multiple threads,
+ *   extern signatures of ffmpeg_opt.c methods called by both ffmpeg and ffprobe added, copied options from
+ *   ffmpeg_opt.c and defined them as inline in execute method)
+ *
+ * 08.2018
+ * --------------------------------------------------------
+ * - fftools_ prefix added to file name and parent headers
+ * - forward_report() method, report_callback function pointer and set_report_callback() setter
+ * method added to forward stats
+ * - forward_report() call added from print_report()
+ * - cancel_operation() method added to trigger sigterm_handler
+ * - (!received_sigterm) validation added inside ifilter_send_eof() to complete cancellation
+ *
+ * 07.2018
+ * --------------------------------------------------------
+ * - main() function renamed as execute()
+ * - exit_program() implemented with setjmp
+ * - extern longjmp_value added to access exit code stored in exit_program()
+ * - cleanup() method added
+ */
+
 #include "config.h"
 #include <ctype.h>
 #include <string.h>
@@ -33,6 +97,9 @@
 #include <stdatomic.h>
 #include <stdint.h>
 
+#include "ffmpegkit_exception.h"
+#include "fftools_opt_common.h"
+
 #if HAVE_IO_H
 #include <io.h>
 #endif
@@ -64,7 +131,6 @@
 #include "libavutil/thread.h"
 #include "libavutil/threadmessage.h"
 #include "libavcodec/mathops.h"
-#include "libavcodec/version.h"
 #include "libavformat/os_support.h"
 
 # include "libavfilter/avfilter.h"
@@ -102,17 +168,14 @@
 
 #include <time.h>
 
-#include "ffmpeg.h"
-#include "cmdutils.h"
-#include "sync_queue.h"
+#include "fftools_ffmpeg.h"
+#include "fftools_cmdutils.h"
+#include "fftools_sync_queue.h"
 
 #include "libavutil/avassert.h"
 
-const char program_name[] = "ffmpeg";
-const int program_birth_year = 2000;
+static __thread FILE *vstats_file;
 
-static FILE *vstats_file;
-
 // optionally attached as opaque_ref to decoded AVFrames
 typedef struct FrameData {
     uint64_t   idx;
@@ -131,31 +194,101 @@
 static int64_t getmaxrss(void);
 static int ifilter_has_all_input_formats(FilterGraph *fg);
 
-static int64_t nb_frames_dup = 0;
-static uint64_t dup_warning = 1000;
-static int64_t nb_frames_drop = 0;
-static int64_t decode_error_stat[2];
-unsigned nb_output_dumped = 0;
+__thread int64_t nb_frames_dup = 0;
+__thread uint64_t dup_warning = 1000;
+__thread int64_t nb_frames_drop = 0;
+__thread int64_t decode_error_stat[2];
+__thread unsigned nb_output_dumped = 0;
 
-static BenchmarkTimeStamps current_time;
-AVIOContext *progress_avio = NULL;
+__thread BenchmarkTimeStamps current_time;
+__thread AVIOContext *progress_avio = NULL;
 
-InputFile   **input_files   = NULL;
-int        nb_input_files   = 0;
+__thread InputFile   **input_files   = NULL;
+__thread int        nb_input_files   = 0;
 
-OutputFile   **output_files   = NULL;
-int         nb_output_files   = 0;
+__thread OutputFile   **output_files   = NULL;
+__thread int         nb_output_files   = 0;
 
-FilterGraph **filtergraphs;
-int        nb_filtergraphs;
+__thread FilterGraph **filtergraphs;
+__thread int        nb_filtergraphs;
+
+__thread int64_t last_time = -1;
+__thread int64_t keyboard_last_time = 0;
+__thread int first_report = 1;
+__thread int qp_histogram[52];
+
+void (*report_callback)(int, float, float, int64_t, double, double, double) = NULL;
+
+extern int opt_map(void *optctx, const char *opt, const char *arg);
+extern int opt_map_channel(void *optctx, const char *opt, const char *arg);
+extern int opt_recording_timestamp(void *optctx, const char *opt, const char *arg);
+extern int opt_data_frames(void *optctx, const char *opt, const char *arg);
+extern int opt_progress(void *optctx, const char *opt, const char *arg);
+extern int opt_target(void *optctx, const char *opt, const char *arg);
+extern int opt_vsync(void *optctx, const char *opt, const char *arg);
+extern int opt_abort_on(void *optctx, const char *opt, const char *arg);
+extern int opt_stats_period(void *optctx, const char *opt, const char *arg);
+extern int opt_qscale(void *optctx, const char *opt, const char *arg);
+extern int opt_profile(void *optctx, const char *opt, const char *arg);
+extern int opt_filter_complex(void *optctx, const char *opt, const char *arg);
+extern int opt_filter_complex_script(void *optctx, const char *opt, const char *arg);
+extern int opt_attach(void *optctx, const char *opt, const char *arg);
+extern int opt_video_frames(void *optctx, const char *opt, const char *arg);
+extern int opt_video_codec(void *optctx, const char *opt, const char *arg);
+extern int opt_sameq(void *optctx, const char *opt, const char *arg);
+extern int opt_timecode(void *optctx, const char *opt, const char *arg);
+extern int opt_vstats_file(void *optctx, const char *opt, const char *arg);
+extern int opt_vstats(void *optctx, const char *opt, const char *arg);
+extern int opt_video_frames(void *optctx, const char *opt, const char *arg);
+extern int opt_old2new(void *optctx, const char *opt, const char *arg);
+extern int opt_streamid(void *optctx, const char *opt, const char *arg);
+extern int opt_bitrate(void *optctx, const char *opt, const char *arg);
+extern int show_hwaccels(void *optctx, const char *opt, const char *arg);
+extern int opt_video_filters(void *optctx, const char *opt, const char *arg);
+extern int opt_audio_frames(void *optctx, const char *opt, const char *arg);
+extern int opt_audio_qscale(void *optctx, const char *opt, const char *arg);
+extern int opt_audio_codec(void *optctx, const char *opt, const char *arg);
+extern int opt_channel_layout(void *optctx, const char *opt, const char *arg);
+extern int opt_preset(void *optctx, const char *opt, const char *arg);
+extern int opt_audio_filters(void *optctx, const char *opt, const char *arg);
+extern int opt_subtitle_codec(void *optctx, const char *opt, const char *arg);
+extern int opt_video_channel(void *optctx, const char *opt, const char *arg);
+extern int opt_video_standard(void *optctx, const char *opt, const char *arg);
+extern int opt_sdp_file(void *optctx, const char *opt, const char *arg);
+#if CONFIG_VAAPI
+extern int opt_vaapi_device(void *optctx, const char *opt, const char *arg);
+#endif
+#if CONFIG_QSV
+extern int opt_qsv_device(void *optctx, const char *opt, const char *arg);
+#endif
+extern int opt_data_codec(void *optctx, const char *opt, const char *arg);
+extern int opt_init_hw_device(void *optctx, const char *opt, const char *arg);
+extern int opt_filter_hw_device(void *optctx, const char *opt, const char *arg);
+extern int opt_filter_threads(void *optctx, const char *opt, const char *arg);
+extern __thread int file_overwrite;
+extern __thread int no_file_overwrite;
+extern __thread int do_psnr;
+extern __thread int ignore_unknown_streams;
+extern __thread int copy_unknown_streams;
+extern __thread int recast_media;
 
 #if HAVE_TERMIOS_H
 
 /* init terminal so that we can grab keys */
-static struct termios oldtty;
-static int restore_tty;
+__thread struct termios oldtty;
+__thread int restore_tty;
 #endif
 
+extern volatile int handleSIGQUIT;
+extern volatile int handleSIGINT;
+extern volatile int handleSIGTERM;
+extern volatile int handleSIGXCPU;
+extern volatile int handleSIGPIPE;
+
+extern __thread long globalSessionId;
+extern void cancelSession(long sessionId);
+extern int cancelRequested(long sessionId);
+
 /* sub2video hack:
    Convert subtitles to video with alpha to insert them in filter graphs.
    This is a temporary solution until libavfilter gets real subtitles support.
@@ -327,26 +460,32 @@
     term_exit_sigsafe();
 }
 
-static volatile int received_sigterm = 0;
-static volatile int received_nb_signals = 0;
-static atomic_int transcode_init_done = ATOMIC_VAR_INIT(0);
-static volatile int ffmpeg_exited = 0;
-int main_return_code = 0;
-static int64_t copy_ts_first_pts = AV_NOPTS_VALUE;
+__thread int received_sigterm = 0;
+__thread int received_nb_signals = 0;
+__thread atomic_int transcode_init_done = ATOMIC_VAR_INIT(0);
+__thread int ffmpeg_exited = 0;
+__thread int main_ffmpeg_return_code = 0;
+__thread int64_t copy_ts_first_pts = AV_NOPTS_VALUE;
+extern __thread int longjmp_value;
+extern __thread int want_sdp;
+struct EncStatsFile;
+extern __thread struct EncStatsFile *enc_stats_files;
+extern __thread int nb_enc_stats_files;
 
 static void
 sigterm_handler(int sig)
 {
-    int ret;
+    // int ret;
     received_sigterm = sig;
     received_nb_signals++;
     term_exit_sigsafe();
-    if(received_nb_signals > 3) {
-        ret = write(2/*STDERR_FILENO*/, "Received > 3 system signals, hard exiting\n",
-                    strlen("Received > 3 system signals, hard exiting\n"));
-        if (ret < 0) { /* Do nothing */ };
-        exit(123);
-    }
+    // FFmpegKit - Hard Exit Disabled
+    // if(received_nb_signals > 3) {
+    //     ret = write(2/*STDERR_FILENO*/, "Received > 3 system signals, hard exiting\n",
+    //                 strlen("Received > 3 system signals, hard exiting\n"));
+    //     if (ret < 0) { /* Do nothing */ };
+    //     exit(123);
+    // }
 }
 
 #if HAVE_SETCONSOLECTRLHANDLER
@@ -395,7 +534,12 @@
 void term_init(void)
 {
 #if defined __linux__
-    struct sigaction action = {0};
+    #if defined  __aarch64__ || defined __amd64__ || defined __x86_64__
+        struct sigaction action = {0};
+    #else
+        struct sigaction action = {{0}};
+    #endif
+
     action.sa_handler = sigterm_handler;
 
     /* block other interrupts while processing this one */
@@ -423,17 +567,27 @@
 
             tcsetattr (0, TCSANOW, &tty);
         }
-        SIGNAL(SIGQUIT, sigterm_handler); /* Quit (POSIX).  */
+        if (handleSIGQUIT == 1) {
+            SIGNAL(SIGQUIT, sigterm_handler); /* Quit (POSIX).  */
+        }
     }
 #endif
 
-    SIGNAL(SIGINT , sigterm_handler); /* Interrupt (ANSI).    */
-    SIGNAL(SIGTERM, sigterm_handler); /* Termination (ANSI).  */
+    if (handleSIGINT == 1) {
+        SIGNAL(SIGINT , sigterm_handler); /* Interrupt (ANSI).    */
+    }
+    if (handleSIGTERM == 1) {
+        SIGNAL(SIGTERM, sigterm_handler); /* Termination (ANSI).  */
+    }
 #ifdef SIGXCPU
-    SIGNAL(SIGXCPU, sigterm_handler);
+    if (handleSIGXCPU == 1) {
+        SIGNAL(SIGXCPU, sigterm_handler);
+    }
 #endif
 #ifdef SIGPIPE
-    signal(SIGPIPE, SIG_IGN); /* Broken pipe (POSIX). */
+    if (handleSIGPIPE == 1) {
+        SIGNAL(SIGPIPE, SIG_IGN); /* Broken pipe (POSIX). */
+    }
 #endif
 #if HAVE_SETCONSOLECTRLHANDLER
     SetConsoleCtrlHandler((PHANDLER_ROUTINE) CtrlHandler, TRUE);
@@ -492,12 +646,14 @@
     return -1;
 }
 
-static int decode_interrupt_cb(void *ctx)
+int decode_interrupt_cb(void *ctx);
+
+int decode_interrupt_cb(void *ctx)
 {
     return received_nb_signals > atomic_load(&transcode_init_done);
 }
 
-const AVIOInterruptCB int_cb = { decode_interrupt_cb, NULL };
+__thread const AVIOInterruptCB int_cb = { decode_interrupt_cb, NULL };
 
 static void ffmpeg_cleanup(int ret)
 {
@@ -576,6 +732,8 @@
     if (received_sigterm) {
         av_log(NULL, AV_LOG_INFO, "Exiting normally, received signal %d.\n",
                (int) received_sigterm);
+    } else if (cancelRequested(globalSessionId)) {
+        av_log(NULL, AV_LOG_INFO, "Exiting normally, received cancel request.\n");
     } else if (ret && atomic_load(&transcode_init_done)) {
         av_log(NULL, AV_LOG_INFO, "Conversion failed!\n");
     }
@@ -1198,7 +1356,7 @@
     }
 }
 
-static enum AVPictureType forced_kf_apply(void *logctx, KeyframeForceCtx *kf,
+enum AVPictureType forced_kf_apply(void *logctx, KeyframeForceCtx *kf,
                                           AVRational tb, const AVFrame *in_picture,
                                           int dup_idx)
 {
@@ -1336,9 +1494,6 @@
 
         in_picture->quality = enc->global_quality;
         in_picture->pict_type = forced_kf_apply(ost, &ost->kf, enc->time_base, in_picture, i);
-
-        if (ost->top_field_first >= 0)
-            in_picture->top_field_first = !!ost->top_field_first;
 
         ret = submit_encode_frame(of, ost, in_picture);
         if (ret == AVERROR_EOF)
@@ -1569,6 +1724,14 @@
     }
 }
 
+static void forward_report(uint64_t frame_number, float fps, float quality, int64_t total_size, int64_t seconds, int64_t microseconds, double bitrate, double speed)
+{
+    // FORWARD DATA
+    if (report_callback != NULL) {
+        report_callback(frame_number, fps, quality, total_size, ((double)seconds*1000) + ((double)microseconds)/1000, bitrate, speed);
+    }
+}
+
 static void print_report(int is_last_report, int64_t timer_start, int64_t cur_time)
 {
     AVBPrint buf, buf_script;
@@ -1577,16 +1740,19 @@
     double bitrate;
     double speed;
     int64_t pts = INT64_MIN + 1;
-    static int64_t last_time = -1;
-    static int first_report = 1;
-    static int qp_histogram[52];
     int hours, mins, secs, us;
     const char *hours_sign;
     int ret;
     float t;
 
+    // FFmpegKit field declarations
+    int local_print_stats = 1;
+    uint64_t frame_number = 0;
+    float fps = 0;
+    float q = 0;
+
     if (!print_stats && !is_last_report && !progress_avio)
-        return;
+        local_print_stats = 0;
 
     if (!is_last_report) {
         if (last_time == -1) {
@@ -1601,40 +1767,44 @@
     t = (cur_time-timer_start) / 1000000.0;
 
     vid = 0;
-    av_bprint_init(&buf, 0, AV_BPRINT_SIZE_AUTOMATIC);
-    av_bprint_init(&buf_script, 0, AV_BPRINT_SIZE_AUTOMATIC);
+    if (local_print_stats) {
+        av_bprint_init(&buf, 0, AV_BPRINT_SIZE_AUTOMATIC);
+        av_bprint_init(&buf_script, 0, AV_BPRINT_SIZE_AUTOMATIC);
+    }
     for (OutputStream *ost = ost_iter(NULL); ost; ost = ost_iter(ost)) {
         const AVCodecContext * const enc = ost->enc_ctx;
-        const float q = enc ? ost->quality / (float) FF_QP2LAMBDA : -1;
+        q = enc ? ost->quality / (float) FF_QP2LAMBDA : -1;
 
-        if (vid && ost->st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO) {
+        if (local_print_stats && vid && ost->st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO) {
             av_bprintf(&buf, "q=%2.1f ", q);
             av_bprintf(&buf_script, "stream_%d_%d_q=%.1f\n",
                        ost->file_index, ost->index, q);
         }
         if (!vid && ost->st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO) {
-            float fps;
-            uint64_t frame_number = atomic_load(&ost->packets_written);
-
+            frame_number = atomic_load(&ost->packets_written);
             fps = t > 1 ? frame_number / t : 0;
-            av_bprintf(&buf, "frame=%5"PRId64" fps=%3.*f q=%3.1f ",
-                     frame_number, fps < 9.95, fps, q);
-            av_bprintf(&buf_script, "frame=%"PRId64"\n", frame_number);
-            av_bprintf(&buf_script, "fps=%.2f\n", fps);
-            av_bprintf(&buf_script, "stream_%d_%d_q=%.1f\n",
-                       ost->file_index, ost->index, q);
-            if (is_last_report)
+            if (local_print_stats) {
+                av_bprintf(&buf, "frame=%5"PRId64" fps=%3.*f q=%3.1f ",
+                         frame_number, fps < 9.95, fps, q);
+                av_bprintf(&buf_script, "frame=%"PRId64"\n", frame_number);
+                av_bprintf(&buf_script, "fps=%.2f\n", fps);
+                av_bprintf(&buf_script, "stream_%d_%d_q=%.1f\n",
+                           ost->file_index, ost->index, q);
+            }
+            if (local_print_stats && is_last_report)
                 av_bprintf(&buf, "L");
             if (qp_hist) {
                 int j;
                 int qp = lrintf(q);
                 if (qp >= 0 && qp < FF_ARRAY_ELEMS(qp_histogram))
                     qp_histogram[qp]++;
-                for (j = 0; j < 32; j++)
-                    av_bprintf(&buf, "%X", av_log2(qp_histogram[j] + 1));
+                if (local_print_stats) {
+                    for (j = 0; j < 32; j++)
+                        av_bprintf(&buf, "%X", av_log2(qp_histogram[j] + 1));
+                }
             }
 
-            if (enc && (enc->flags & AV_CODEC_FLAG_PSNR) &&
+            if (local_print_stats && enc && (enc->flags & AV_CODEC_FLAG_PSNR) &&
                 (ost->pict_type != AV_PICTURE_TYPE_NONE || is_last_report)) {
                 int j;
                 double error, error_sum = 0;
@@ -1692,78 +1862,81 @@
     bitrate = pts && total_size >= 0 ? total_size * 8 / (pts / 1000.0) : -1;
     speed = t != 0.0 ? (double)pts / AV_TIME_BASE / t : -1;
 
-    if (total_size < 0) av_bprintf(&buf, "size=N/A time=");
-    else                av_bprintf(&buf, "size=%8.0fkB time=", total_size / 1024.0);
-    if (pts == AV_NOPTS_VALUE) {
-        av_bprintf(&buf, "N/A ");
-    } else {
-        av_bprintf(&buf, "%s%02d:%02d:%02d.%02d ",
-                   hours_sign, hours, mins, secs, (100 * us) / AV_TIME_BASE);
-    }
+    // FFmpegKit forward report
+    forward_report(frame_number, fps, q, total_size, secs, us, bitrate, speed);
 
-    if (bitrate < 0) {
-        av_bprintf(&buf, "bitrate=N/A");
-        av_bprintf(&buf_script, "bitrate=N/A\n");
-    }else{
-        av_bprintf(&buf, "bitrate=%6.1fkbits/s", bitrate);
-        av_bprintf(&buf_script, "bitrate=%6.1fkbits/s\n", bitrate);
-    }
+    if (local_print_stats) {
+        if (total_size < 0) av_bprintf(&buf, "size=N/A time=");
+        else                av_bprintf(&buf, "size=%8.0fkB time=", total_size / 1024.0);
+        if (pts == AV_NOPTS_VALUE) {
+            av_bprintf(&buf, "N/A ");
+        } else {
+            av_bprintf(&buf, "%s%02d:%02d:%02d.%02d ",
+                       hours_sign, hours, mins, secs, (100 * us) / AV_TIME_BASE);
+        }
 
-    if (total_size < 0) av_bprintf(&buf_script, "total_size=N/A\n");
-    else                av_bprintf(&buf_script, "total_size=%"PRId64"\n", total_size);
-    if (pts == AV_NOPTS_VALUE) {
-        av_bprintf(&buf_script, "out_time_us=N/A\n");
-        av_bprintf(&buf_script, "out_time_ms=N/A\n");
-        av_bprintf(&buf_script, "out_time=N/A\n");
-    } else {
-        av_bprintf(&buf_script, "out_time_us=%"PRId64"\n", pts);
-        av_bprintf(&buf_script, "out_time_ms=%"PRId64"\n", pts);
-        av_bprintf(&buf_script, "out_time=%s%02d:%02d:%02d.%06d\n",
-                   hours_sign, hours, mins, secs, us);
-    }
+        if (bitrate < 0) {
+            av_bprintf(&buf, "bitrate=N/A");
+            av_bprintf(&buf_script, "bitrate=N/A\n");
+        }else{
+            av_bprintf(&buf, "bitrate=%6.1fkbits/s", bitrate);
+            av_bprintf(&buf_script, "bitrate=%6.1fkbits/s\n", bitrate);
+        }
 
-    if (nb_frames_dup || nb_frames_drop)
-        av_bprintf(&buf, " dup=%"PRId64" drop=%"PRId64, nb_frames_dup, nb_frames_drop);
-    av_bprintf(&buf_script, "dup_frames=%"PRId64"\n", nb_frames_dup);
-    av_bprintf(&buf_script, "drop_frames=%"PRId64"\n", nb_frames_drop);
-
-    if (speed < 0) {
-        av_bprintf(&buf, " speed=N/A");
-        av_bprintf(&buf_script, "speed=N/A\n");
-    } else {
-        av_bprintf(&buf, " speed=%4.3gx", speed);
-        av_bprintf(&buf_script, "speed=%4.3gx\n", speed);
-    }
-
-    if (print_stats || is_last_report) {
-        const char end = is_last_report ? '\n' : '\r';
-        if (print_stats==1 && AV_LOG_INFO > av_log_get_level()) {
-            fprintf(stderr, "%s    %c", buf.str, end);
-        } else
-            av_log(NULL, AV_LOG_INFO, "%s    %c", buf.str, end);
+        if (total_size < 0) av_bprintf(&buf_script, "total_size=N/A\n");
+        else                av_bprintf(&buf_script, "total_size=%"PRId64"\n", total_size);
+        if (pts == AV_NOPTS_VALUE) {
+            av_bprintf(&buf_script, "out_time_us=N/A\n");
+            av_bprintf(&buf_script, "out_time_ms=N/A\n");
+            av_bprintf(&buf_script, "out_time=N/A\n");
+        } else {
+            av_bprintf(&buf_script, "out_time_us=%"PRId64"\n", pts);
+            av_bprintf(&buf_script, "out_time_ms=%"PRId64"\n", pts);
+            av_bprintf(&buf_script, "out_time=%s%02d:%02d:%02d.%06d\n",
+                       hours_sign, hours, mins, secs, us);
+        }
 
-        fflush(stderr);
-    }
-    av_bprint_finalize(&buf, NULL);
+        if (nb_frames_dup || nb_frames_drop)
+            av_bprintf(&buf, " dup=%"PRId64" drop=%"PRId64, nb_frames_dup, nb_frames_drop);
+        av_bprintf(&buf_script, "dup_frames=%"PRId64"\n", nb_frames_dup);
+        av_bprintf(&buf_script, "drop_frames=%"PRId64"\n", nb_frames_drop);
 
-    if (progress_avio) {
-        av_bprintf(&buf_script, "progress=%s\n",
-                   is_last_report ? "end" : "continue");
-        avio_write(progress_avio, buf_script.str,
-                   FFMIN(buf_script.len, buf_script.size - 1));
-        avio_flush(progress_avio);
-        av_bprint_finalize(&buf_script, NULL);
-        if (is_last_report) {
-            if ((ret = avio_closep(&progress_avio)) < 0)
-                av_log(NULL, AV_LOG_ERROR,
-                       "Error closing progress log, loss of information possible: %s\n", av_err2str(ret));
+        if (speed < 0) {
+            av_bprintf(&buf, " speed=N/A");
+            av_bprintf(&buf_script, "speed=N/A\n");
+        } else {
+            av_bprintf(&buf, " speed=%4.3gx", speed);
+            av_bprintf(&buf_script, "speed=%4.3gx\n", speed);
         }
-    }
 
-    first_report = 0;
+        if (print_stats || is_last_report) {
+            const char end = is_last_report ? '\n' : '\r';
+            if (print_stats==1 && AV_LOG_INFO > av_log_get_level()) {
+                av_log(NULL, AV_LOG_STDERR, "%s    %c", buf.str, end);
+            } else
+                av_log(NULL, AV_LOG_INFO, "%s    %c", buf.str, end);
+        }
+        av_bprint_finalize(&buf, NULL);
 
-    if (is_last_report)
-        print_final_stats(total_size);
+        if (progress_avio) {
+            av_bprintf(&buf_script, "progress=%s\n",
+                       is_last_report ? "end" : "continue");
+            avio_write(progress_avio, buf_script.str,
+                       FFMIN(buf_script.len, buf_script.size - 1));
+            avio_flush(progress_avio);
+            av_bprint_finalize(&buf_script, NULL);
+            if (is_last_report) {
+                if ((ret = avio_closep(&progress_avio)) < 0)
+                    av_log(NULL, AV_LOG_ERROR,
+                           "Error closing progress log, loss of information possible: %s\n", av_err2str(ret));
+            }
+        }
+
+        first_report = 0;
+
+        if (is_last_report)
+            print_final_stats(total_size);
+    }
 }
 
 static int ifilter_parameters_from_codecpar(InputFilter *ifilter, AVCodecParameters *par)
@@ -2008,7 +2181,7 @@
         (ifilter->hw_frames_ctx && ifilter->hw_frames_ctx->data != frame->hw_frames_ctx->data))
         need_reinit = 1;
 
-    if (sd = av_frame_get_side_data(frame, AV_FRAME_DATA_DISPLAYMATRIX)) {
+    if ((sd = av_frame_get_side_data(frame, AV_FRAME_DATA_DISPLAYMATRIX))) {
         if (!ifilter->displaymatrix || memcmp(sd->data, ifilter->displaymatrix, sizeof(int32_t) * 9))
             need_reinit = 1;
     } else if (ifilter->displaymatrix)
@@ -2059,12 +2232,16 @@
 
 static int ifilter_send_eof(InputFilter *ifilter, int64_t pts)
 {
-    int ret;
+    int ret = 0;
 
     ifilter->eof = 1;
 
     if (ifilter->filter) {
-        ret = av_buffersrc_close(ifilter->filter, pts, AV_BUFFERSRC_FLAG_PUSH);
+
+        /* THIS VALIDATION IS REQUIRED TO COMPLETE CANCELLATION */
+        if (!received_sigterm && !cancelRequested(globalSessionId)) {
+            ret = av_buffersrc_close(ifilter->filter, pts, AV_BUFFERSRC_FLAG_PUSH);
+        }
         if (ret < 0)
             return ret;
     } else {
@@ -3517,13 +3694,12 @@
 static int check_keyboard_interaction(int64_t cur_time)
 {
     int i, ret, key;
-    static int64_t last_time;
     if (received_nb_signals)
         return AVERROR_EXIT;
     /* read_key() returns 0 on EOF */
-    if (cur_time - last_time >= 100000) {
+    if(cur_time - keyboard_last_time >= 100000){
         key =  read_key();
-        last_time = cur_time;
+        keyboard_last_time = cur_time;
     }else
         key = -1;
     if (key == 'q') {
@@ -3537,7 +3713,7 @@
         char buf[4096], target[64], command[256], arg[256] = {0};
         double time;
         int k, n = 0;
-        fprintf(stderr, "\nEnter command: <target>|all <time>|-1 <command>[ <argument>]\n");
+        av_log(NULL, AV_LOG_STDERR, "\nEnter command: <target>|all <time>|-1 <command>[ <argument>]\n");
         i = 0;
         set_tty_echo(1);
         while ((k = read_key()) != '\n' && k != '\r' && i < sizeof(buf)-1)
@@ -3545,7 +3721,7 @@
                 buf[i++] = k;
         buf[i] = 0;
         set_tty_echo(0);
-        fprintf(stderr, "\n");
+        av_log(NULL, AV_LOG_STDERR, "\n");
         if (k > 0 &&
             (n = sscanf(buf, "%63[^ ] %lf %255[^ ] %255[^\n]", target, &time, command, arg)) >= 3) {
             av_log(NULL, AV_LOG_DEBUG, "Processing command target:%s time:%f command:%s arg:%s",
@@ -3556,14 +3732,14 @@
                     if (time < 0) {
                         ret = avfilter_graph_send_command(fg->graph, target, command, arg, buf, sizeof(buf),
                                                           key == 'c' ? AVFILTER_CMD_FLAG_ONE : 0);
-                        fprintf(stderr, "Command reply for stream %d: ret:%d res:\n%s", i, ret, buf);
+                        av_log(NULL, AV_LOG_STDERR, "Command reply for stream %d: ret:%d res:\n%s", i, ret, buf);
                     } else if (key == 'c') {
-                        fprintf(stderr, "Queuing commands only on filters supporting the specific command is unsupported\n");
+                        av_log(NULL, AV_LOG_STDERR, "Queuing commands only on filters supporting the specific command is unsupported\n");
                         ret = AVERROR_PATCHWELCOME;
                     } else {
                         ret = avfilter_graph_queue_command(fg->graph, target, command, arg, 0, time);
                         if (ret < 0)
-                            fprintf(stderr, "Queuing command failed with error %s\n", av_err2str(ret));
+                            av_log(NULL, AV_LOG_STDERR, "Queuing command failed with error %s\n", av_err2str(ret));
                     }
                 }
             }
@@ -3594,9 +3770,9 @@
                     buf[i++] = k;
             buf[i] = 0;
             set_tty_echo(0);
-            fprintf(stderr, "\n");
+            av_log(NULL, AV_LOG_STDERR, "\n");
             if (k <= 0 || sscanf(buf, "%d", &debug)!=1)
-                fprintf(stderr,"error parsing debug value\n");
+                av_log(NULL, AV_LOG_STDERR,"error parsing debug value\n");
         }
         for (InputStream *ist = ist_iter(NULL); ist; ist = ist_iter(ist))
             ist->dec_ctx->debug = debug;
@@ -3605,10 +3781,10 @@
                 ost->enc_ctx->debug = debug;
         }
         if(debug) av_log_set_level(AV_LOG_DEBUG);
-        fprintf(stderr,"debug=%d\n", debug);
+        av_log(NULL, AV_LOG_STDERR,"debug=%d\n", debug);
     }
     if (key == '?'){
-        fprintf(stderr, "key    function\n"
+        av_log(NULL, AV_LOG_STDERR, "key    function\n"
                         "?      show this help\n"
                         "+      increase verbosity\n"
                         "-      decrease verbosity\n"
@@ -4025,7 +4201,7 @@
 
     timer_start = av_gettime_relative();
 
-    while (!received_sigterm) {
+    while (!received_sigterm && !cancelRequested(globalSessionId)) {
         int64_t cur_time= av_gettime_relative();
 
         /* if 'q' pressed, exits */
@@ -4138,62 +4314,599 @@
 #endif
 }
 
-int main(int argc, char **argv)
-{
-    int ret;
-    BenchmarkTimeStamps ti;
+void ffmpeg_var_cleanup() {
+    received_sigterm = 0;
+    received_nb_signals = 0;
+    transcode_init_done = ATOMIC_VAR_INIT(0);
+    ffmpeg_exited = 0;
+    main_ffmpeg_return_code = 0;
+    copy_ts_first_pts = AV_NOPTS_VALUE;
+    longjmp_value = 0;
+    want_sdp = 1;
+    enc_stats_files = NULL;
+    nb_enc_stats_files = 0;
 
-    init_dynload();
+    vstats_file = NULL;
 
-    register_exit(ffmpeg_cleanup);
+    nb_frames_dup = 0;
+    dup_warning = 1000;
+    nb_frames_drop = 0;
+    decode_error_stat[0] = 0;
+    decode_error_stat[1] = 0;
+    nb_output_dumped = 0;
 
-    setvbuf(stderr,NULL,_IONBF,0); /* win32 runtime needs this */
+    progress_avio = NULL;
 
-    av_log_set_flags(AV_LOG_SKIP_REPEATED);
-    parse_loglevel(argc, argv, options);
+    input_files = NULL;
+    nb_input_files = 0;
 
-#if CONFIG_AVDEVICE
-    avdevice_register_all();
-#endif
-    avformat_network_init();
+    output_files = NULL;
+    nb_output_files = 0;
 
-    show_banner(argc, argv, options);
+    filtergraphs = NULL;
+    nb_filtergraphs = 0;
 
-    /* parse options and open all input/output files */
-    ret = ffmpeg_parse_options(argc, argv);
-    if (ret < 0)
-        exit_program(1);
+    last_time = -1;
+    keyboard_last_time = 0;
+    first_report = 1;
+    for(int i = 0; i < FF_ARRAY_ELEMS(qp_histogram); i++) {
+        qp_histogram[i] = 0;
+    }
+}
 
-    if (nb_output_files <= 0 && nb_input_files == 0) {
-        show_usage();
-        av_log(NULL, AV_LOG_WARNING, "Use -h to get full help or, even better, run 'man %s'\n", program_name);
-        exit_program(1);
+void set_report_callback(void (*callback)(int, float, float, int64_t, double, double, double))
+{
+    report_callback = callback;
+}
+
+void cancel_operation(long id)
+{
+    if (id == 0) {
+        sigterm_handler(SIGINT);
+    } else {
+        cancelSession(id);
     }
+}
 
-    /* file converter / grab */
-    if (nb_output_files <= 0) {
-        av_log(NULL, AV_LOG_FATAL, "At least one output file must be specified\n");
-        exit_program(1);
-    }
+__thread OptionDef *ffmpeg_options = NULL;
+
+int ffmpeg_execute(int argc, char **argv)
+{
+    char _program_name[] = "ffmpeg";
+    program_name = (char*)&_program_name;
+    program_birth_year = 2000;
+
+    #define OFFSET(x) offsetof(OptionsContext, x)
+    OptionDef options[] = {
 
-    current_time = ti = get_benchmark_time_stamps();
-    if (transcode() < 0)
-        exit_program(1);
-    if (do_benchmark) {
-        int64_t utime, stime, rtime;
-        current_time = get_benchmark_time_stamps();
-        utime = current_time.user_usec - ti.user_usec;
-        stime = current_time.sys_usec  - ti.sys_usec;
-        rtime = current_time.real_usec - ti.real_usec;
-        av_log(NULL, AV_LOG_INFO,
-               "bench: utime=%0.3fs stime=%0.3fs rtime=%0.3fs\n",
-               utime / 1000000.0, stime / 1000000.0, rtime / 1000000.0);
-    }
-    av_log(NULL, AV_LOG_DEBUG, "%"PRIu64" frames successfully decoded, %"PRIu64" decoding errors\n",
-           decode_error_stat[0], decode_error_stat[1]);
-    if ((decode_error_stat[0] + decode_error_stat[1]) * max_error_rate < decode_error_stat[1])
-        exit_program(69);
+        /* main options */
+        { "L",           OPT_EXIT,             { .func_arg = show_license },     "show license" },
+        { "h",           OPT_EXIT,             { .func_arg = show_help },        "show help", "topic" },
+        { "?",           OPT_EXIT,             { .func_arg = show_help },        "show help", "topic" },
+        { "help",        OPT_EXIT,             { .func_arg = show_help },        "show help", "topic" },
+        { "-help",       OPT_EXIT,             { .func_arg = show_help },        "show help", "topic" },
+        { "version",     OPT_EXIT,             { .func_arg = show_version },     "show version" },
+        { "buildconf",   OPT_EXIT,             { .func_arg = show_buildconf },   "show build configuration" },
+        { "formats",     OPT_EXIT,             { .func_arg = show_formats },     "show available formats" },
+        { "muxers",      OPT_EXIT,             { .func_arg = show_muxers },      "show available muxers" },
+        { "demuxers",    OPT_EXIT,             { .func_arg = show_demuxers },    "show available demuxers" },
+        { "devices",     OPT_EXIT,             { .func_arg = show_devices },     "show available devices" },
+        { "codecs",      OPT_EXIT,             { .func_arg = show_codecs },      "show available codecs" },
+        { "decoders",    OPT_EXIT,             { .func_arg = show_decoders },    "show available decoders" },
+        { "encoders",    OPT_EXIT,             { .func_arg = show_encoders },    "show available encoders" },
+        { "bsfs",        OPT_EXIT,             { .func_arg = show_bsfs },        "show available bit stream filters" },
+        { "protocols",   OPT_EXIT,             { .func_arg = show_protocols },   "show available protocols" },
+        { "filters",     OPT_EXIT,             { .func_arg = show_filters },     "show available filters" },
+        { "pix_fmts",    OPT_EXIT,             { .func_arg = show_pix_fmts },    "show available pixel formats" },
+        { "layouts",     OPT_EXIT,             { .func_arg = show_layouts },     "show standard channel layouts" },
+        { "sample_fmts", OPT_EXIT,             { .func_arg = show_sample_fmts }, "show available audio sample formats" },
+        { "dispositions", OPT_EXIT,            { .func_arg = show_dispositions}, "show available stream dispositions" },
+        { "colors",      OPT_EXIT,             { .func_arg = show_colors },      "show available color names" },
+        { "loglevel",    HAS_ARG,              { .func_arg = opt_loglevel },     "set logging level", "loglevel" },
+        { "v",           HAS_ARG,              { .func_arg = opt_loglevel },     "set logging level", "loglevel" },
+        { "report",      0,                    { .func_arg = opt_report },       "generate a report" },
+        { "max_alloc",   HAS_ARG,              { .func_arg = opt_max_alloc },    "set maximum size of a single allocated block", "bytes" },
+        { "cpuflags",    HAS_ARG | OPT_EXPERT, { .func_arg = opt_cpuflags },     "force specific cpu flags", "flags" },
+        { "cpucount",    HAS_ARG | OPT_EXPERT, { .func_arg = opt_cpucount },     "force specific cpu count", "count" },
+        { "hide_banner", OPT_BOOL | OPT_EXPERT, {&hide_banner},     "do not show program banner", "hide_banner" },
 
-    exit_program(received_nb_signals ? 255 : main_return_code);
-    return main_return_code;
+        #if CONFIG_AVDEVICE
+            { "sources"    , OPT_EXIT | HAS_ARG, { .func_arg = show_sources },
+              "list sources of the input device", "device" },
+            { "sinks"      , OPT_EXIT | HAS_ARG, { .func_arg = show_sinks },
+              "list sinks of the output device", "device" },
+        #endif
+
+        { "f",              HAS_ARG | OPT_STRING | OPT_OFFSET |
+                            OPT_INPUT | OPT_OUTPUT,                      { .off       = OFFSET(format) },
+            "force format", "fmt" },
+        { "y",              OPT_BOOL,                                    {              &file_overwrite },
+            "overwrite output files" },
+        { "n",              OPT_BOOL,                                    {              &no_file_overwrite },
+            "never overwrite output files" },
+        { "ignore_unknown", OPT_BOOL,                                    {              &ignore_unknown_streams },
+            "Ignore unknown stream types" },
+        { "copy_unknown",   OPT_BOOL | OPT_EXPERT,                       {              &copy_unknown_streams },
+            "Copy unknown stream types" },
+        { "recast_media",   OPT_BOOL | OPT_EXPERT,                       {              &recast_media },
+            "allow recasting stream type in order to force a decoder of different media type" },
+        { "c",              HAS_ARG | OPT_STRING | OPT_SPEC |
+                            OPT_INPUT | OPT_OUTPUT,                      { .off       = OFFSET(codec_names) },
+            "codec name", "codec" },
+        { "codec",          HAS_ARG | OPT_STRING | OPT_SPEC |
+                            OPT_INPUT | OPT_OUTPUT,                      { .off       = OFFSET(codec_names) },
+            "codec name", "codec" },
+        { "pre",            HAS_ARG | OPT_STRING | OPT_SPEC |
+                            OPT_OUTPUT,                                  { .off       = OFFSET(presets) },
+            "preset name", "preset" },
+        { "map",            HAS_ARG | OPT_EXPERT | OPT_PERFILE |
+                            OPT_OUTPUT,                                  { .func_arg = opt_map },
+            "set input stream mapping",
+            "[-]input_file_id[:stream_specifier][,sync_file_id[:stream_specifier]]" },
+    #if FFMPEG_OPT_MAP_CHANNEL
+        { "map_channel",    HAS_ARG | OPT_EXPERT | OPT_PERFILE | OPT_OUTPUT, { .func_arg = opt_map_channel },
+            "map an audio channel from one stream to another (deprecated)", "file.stream.channel[:syncfile.syncstream]" },
+    #endif
+        { "map_metadata",   HAS_ARG | OPT_STRING | OPT_SPEC |
+                            OPT_OUTPUT,                                  { .off       = OFFSET(metadata_map) },
+            "set metadata information of outfile from infile",
+            "outfile[,metadata]:infile[,metadata]" },
+        { "map_chapters",   HAS_ARG | OPT_INT | OPT_EXPERT | OPT_OFFSET |
+                            OPT_OUTPUT,                                  { .off = OFFSET(chapters_input_file) },
+            "set chapters mapping", "input_file_index" },
+        { "t",              HAS_ARG | OPT_TIME | OPT_OFFSET |
+                            OPT_INPUT | OPT_OUTPUT,                      { .off = OFFSET(recording_time) },
+            "record or transcode \"duration\" seconds of audio/video",
+            "duration" },
+        { "to",             HAS_ARG | OPT_TIME | OPT_OFFSET | OPT_INPUT | OPT_OUTPUT,  { .off = OFFSET(stop_time) },
+            "record or transcode stop time", "time_stop" },
+        { "fs",             HAS_ARG | OPT_INT64 | OPT_OFFSET | OPT_OUTPUT, { .off = OFFSET(limit_filesize) },
+            "set the limit file size in bytes", "limit_size" },
+        { "ss",             HAS_ARG | OPT_TIME | OPT_OFFSET |
+                            OPT_INPUT | OPT_OUTPUT,                      { .off = OFFSET(start_time) },
+            "set the start time offset", "time_off" },
+        { "sseof",          HAS_ARG | OPT_TIME | OPT_OFFSET |
+                            OPT_INPUT,                                   { .off = OFFSET(start_time_eof) },
+            "set the start time offset relative to EOF", "time_off" },
+        { "seek_timestamp", HAS_ARG | OPT_INT | OPT_OFFSET |
+                            OPT_INPUT,                                   { .off = OFFSET(seek_timestamp) },
+            "enable/disable seeking by timestamp with -ss" },
+        { "accurate_seek",  OPT_BOOL | OPT_OFFSET | OPT_EXPERT |
+                            OPT_INPUT,                                   { .off = OFFSET(accurate_seek) },
+            "enable/disable accurate seeking with -ss" },
+        { "isync",          HAS_ARG | OPT_INT | OPT_OFFSET |
+                            OPT_EXPERT | OPT_INPUT,                      { .off = OFFSET(input_sync_ref) },
+            "Indicate the input index for sync reference", "sync ref" },
+        { "itsoffset",      HAS_ARG | OPT_TIME | OPT_OFFSET |
+                            OPT_EXPERT | OPT_INPUT,                      { .off = OFFSET(input_ts_offset) },
+            "set the input ts offset", "time_off" },
+        { "itsscale",       HAS_ARG | OPT_DOUBLE | OPT_SPEC |
+                            OPT_EXPERT | OPT_INPUT,                      { .off = OFFSET(ts_scale) },
+            "set the input ts scale", "scale" },
+        { "timestamp",      HAS_ARG | OPT_PERFILE | OPT_OUTPUT,          { .func_arg = opt_recording_timestamp },
+            "set the recording timestamp ('now' to set the current time)", "time" },
+        { "metadata",       HAS_ARG | OPT_STRING | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(metadata) },
+            "add metadata", "string=string" },
+        { "program",        HAS_ARG | OPT_STRING | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(program) },
+            "add program with specified streams", "title=string:st=number..." },
+        { "dframes",        HAS_ARG | OPT_PERFILE | OPT_EXPERT |
+                            OPT_OUTPUT,                                  { .func_arg = opt_data_frames },
+            "set the number of data frames to output", "number" },
+        { "benchmark",      OPT_BOOL | OPT_EXPERT,                       { &do_benchmark },
+            "add timings for benchmarking" },
+        { "benchmark_all",  OPT_BOOL | OPT_EXPERT,                       { &do_benchmark_all },
+          "add timings for each task" },
+        { "progress",       HAS_ARG | OPT_EXPERT,                        { .func_arg = opt_progress },
+          "write program-readable progress information", "url" },
+        { "stdin",          OPT_BOOL | OPT_EXPERT,                       { &stdin_interaction },
+          "enable or disable interaction on standard input" },
+        { "timelimit",      HAS_ARG | OPT_EXPERT,                        { .func_arg = opt_timelimit },
+            "set max runtime in seconds in CPU user time", "limit" },
+        { "dump",           OPT_BOOL | OPT_EXPERT,                       { &do_pkt_dump },
+            "dump each input packet" },
+        { "hex",            OPT_BOOL | OPT_EXPERT,                       { &do_hex_dump },
+            "when dumping packets, also dump the payload" },
+        { "re",             OPT_BOOL | OPT_EXPERT | OPT_OFFSET |
+                            OPT_INPUT,                                   { .off = OFFSET(rate_emu) },
+            "read input at native frame rate; equivalent to -readrate 1", "" },
+        { "readrate",       HAS_ARG | OPT_FLOAT | OPT_OFFSET |
+                            OPT_EXPERT | OPT_INPUT,                      { .off = OFFSET(readrate) },
+            "read input at specified rate", "speed" },
+        { "target",         HAS_ARG | OPT_PERFILE | OPT_OUTPUT,          { .func_arg = opt_target },
+            "specify target file type (\"vcd\", \"svcd\", \"dvd\", \"dv\" or \"dv50\" "
+            "with optional prefixes \"pal-\", \"ntsc-\" or \"film-\")", "type" },
+        { "vsync",          HAS_ARG | OPT_EXPERT,                        { .func_arg = opt_vsync },
+            "set video sync method globally; deprecated, use -fps_mode", "" },
+        { "frame_drop_threshold", HAS_ARG | OPT_FLOAT | OPT_EXPERT,      { &frame_drop_threshold },
+            "frame drop threshold", "" },
+        { "adrift_threshold", HAS_ARG | OPT_FLOAT | OPT_EXPERT,          { &audio_drift_threshold },
+            "audio drift threshold", "threshold" },
+        { "copyts",         OPT_BOOL | OPT_EXPERT,                       { &copy_ts },
+            "copy timestamps" },
+        { "start_at_zero",  OPT_BOOL | OPT_EXPERT,                       { &start_at_zero },
+            "shift input timestamps to start at 0 when using copyts" },
+        { "copytb",         HAS_ARG | OPT_INT | OPT_EXPERT,              { &copy_tb },
+            "copy input stream time base when stream copying", "mode" },
+        { "shortest",       OPT_BOOL | OPT_EXPERT | OPT_OFFSET |
+                            OPT_OUTPUT,                                  { .off = OFFSET(shortest) },
+            "finish encoding within shortest input" },
+        { "shortest_buf_duration", HAS_ARG | OPT_FLOAT | OPT_EXPERT | OPT_OFFSET | OPT_OUTPUT, { .off = OFFSET(shortest_buf_duration) },
+            "maximum buffering duration (in seconds) for the -shortest option" },
+        { "bitexact",       OPT_BOOL | OPT_EXPERT | OPT_OFFSET |
+                            OPT_OUTPUT | OPT_INPUT,                      { .off = OFFSET(bitexact) },
+            "bitexact mode" },
+        { "apad",           OPT_STRING | HAS_ARG | OPT_SPEC |
+                            OPT_OUTPUT,                                  { .off = OFFSET(apad) },
+            "audio pad", "" },
+        { "dts_delta_threshold", HAS_ARG | OPT_FLOAT | OPT_EXPERT,       { &dts_delta_threshold },
+            "timestamp discontinuity delta threshold", "threshold" },
+        { "dts_error_threshold", HAS_ARG | OPT_FLOAT | OPT_EXPERT,       { &dts_error_threshold },
+            "timestamp error delta threshold", "threshold" },
+        { "xerror",         OPT_BOOL | OPT_EXPERT,                       { &exit_on_error },
+            "exit on error", "error" },
+        { "abort_on",       HAS_ARG | OPT_EXPERT,                        { .func_arg = opt_abort_on },
+            "abort on the specified condition flags", "flags" },
+        { "copyinkf",       OPT_BOOL | OPT_EXPERT | OPT_SPEC |
+                            OPT_OUTPUT,                                  { .off = OFFSET(copy_initial_nonkeyframes) },
+            "copy initial non-keyframes" },
+        { "copypriorss",    OPT_INT | HAS_ARG | OPT_EXPERT | OPT_SPEC | OPT_OUTPUT,   { .off = OFFSET(copy_prior_start) },
+            "copy or discard frames before start time" },
+        { "frames",         OPT_INT64 | HAS_ARG | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(max_frames) },
+            "set the number of frames to output", "number" },
+        { "tag",            OPT_STRING | HAS_ARG | OPT_SPEC |
+                            OPT_EXPERT | OPT_OUTPUT | OPT_INPUT,         { .off = OFFSET(codec_tags) },
+            "force codec tag/fourcc", "fourcc/tag" },
+        { "q",              HAS_ARG | OPT_EXPERT | OPT_DOUBLE |
+                            OPT_SPEC | OPT_OUTPUT,                       { .off = OFFSET(qscale) },
+            "use fixed quality scale (VBR)", "q" },
+        { "qscale",         HAS_ARG | OPT_EXPERT | OPT_PERFILE |
+                            OPT_OUTPUT,                                  { .func_arg = opt_qscale },
+            "use fixed quality scale (VBR)", "q" },
+        { "profile",        HAS_ARG | OPT_EXPERT | OPT_PERFILE | OPT_OUTPUT, { .func_arg = opt_profile },
+            "set profile", "profile" },
+        { "filter",         HAS_ARG | OPT_STRING | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(filters) },
+            "set stream filtergraph", "filter_graph" },
+        { "filter_threads", HAS_ARG,                                     { .func_arg = opt_filter_threads },
+            "number of non-complex filter threads" },
+        { "filter_script",  HAS_ARG | OPT_STRING | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(filter_scripts) },
+            "read stream filtergraph description from a file", "filename" },
+        { "reinit_filter",  HAS_ARG | OPT_INT | OPT_SPEC | OPT_INPUT,    { .off = OFFSET(reinit_filters) },
+            "reinit filtergraph on input parameter changes", "" },
+        { "filter_complex", HAS_ARG | OPT_EXPERT,                        { .func_arg = opt_filter_complex },
+            "create a complex filtergraph", "graph_description" },
+        { "filter_complex_threads", HAS_ARG | OPT_INT,                   { &filter_complex_nbthreads },
+            "number of threads for -filter_complex" },
+        { "lavfi",          HAS_ARG | OPT_EXPERT,                        { .func_arg = opt_filter_complex },
+            "create a complex filtergraph", "graph_description" },
+        { "filter_complex_script", HAS_ARG | OPT_EXPERT,                 { .func_arg = opt_filter_complex_script },
+            "read complex filtergraph description from a file", "filename" },
+        { "auto_conversion_filters", OPT_BOOL | OPT_EXPERT,              { &auto_conversion_filters },
+            "enable automatic conversion filters globally" },
+        { "stats",          OPT_BOOL,                                    { &print_stats },
+            "print progress report during encoding", },
+        { "stats_period",    HAS_ARG | OPT_EXPERT,                       { .func_arg = opt_stats_period },
+            "set the period at which ffmpeg updates stats and -progress output", "time" },
+        { "attach",         HAS_ARG | OPT_PERFILE | OPT_EXPERT |
+                            OPT_OUTPUT,                                  { .func_arg = opt_attach },
+            "add an attachment to the output file", "filename" },
+        { "dump_attachment", HAS_ARG | OPT_STRING | OPT_SPEC |
+                             OPT_EXPERT | OPT_INPUT,                     { .off = OFFSET(dump_attachment) },
+            "extract an attachment into a file", "filename" },
+        { "stream_loop", OPT_INT | HAS_ARG | OPT_EXPERT | OPT_INPUT |
+                            OPT_OFFSET,                                  { .off = OFFSET(loop) }, "set number of times input stream shall be looped", "loop count" },
+        { "debug_ts",       OPT_BOOL | OPT_EXPERT,                       { &debug_ts },
+            "print timestamp debugging info" },
+        { "max_error_rate",  HAS_ARG | OPT_FLOAT,                        { &max_error_rate },
+            "ratio of decoding errors (0.0: no errors, 1.0: 100% errors) above which ffmpeg returns an error instead of success.", "maximum error rate" },
+        { "discard",        OPT_STRING | HAS_ARG | OPT_SPEC |
+                            OPT_INPUT,                                   { .off = OFFSET(discard) },
+            "discard", "" },
+        { "disposition",    OPT_STRING | HAS_ARG | OPT_SPEC |
+                            OPT_OUTPUT,                                  { .off = OFFSET(disposition) },
+            "disposition", "" },
+        { "thread_queue_size", HAS_ARG | OPT_INT | OPT_OFFSET | OPT_EXPERT | OPT_INPUT | OPT_OUTPUT,
+                                                                         { .off = OFFSET(thread_queue_size) },
+            "set the maximum number of queued packets from the demuxer" },
+        { "find_stream_info", OPT_BOOL | OPT_INPUT | OPT_EXPERT | OPT_OFFSET, { .off = OFFSET(find_stream_info) },
+            "read and decode the streams to fill missing information with heuristics" },
+        { "bits_per_raw_sample", OPT_INT | HAS_ARG | OPT_EXPERT | OPT_SPEC | OPT_OUTPUT,
+            { .off = OFFSET(bits_per_raw_sample) },
+            "set the number of bits per raw sample", "number" },
+
+        { "stats_enc_pre",      HAS_ARG | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT | OPT_STRING, { .off = OFFSET(enc_stats_pre)      },
+            "write encoding stats before encoding" },
+        { "stats_enc_post",     HAS_ARG | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT | OPT_STRING, { .off = OFFSET(enc_stats_post)     },
+            "write encoding stats after encoding" },
+        { "stats_mux_pre",      HAS_ARG | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT | OPT_STRING, { .off = OFFSET(mux_stats)          },
+            "write packets stats before muxing" },
+        { "stats_enc_pre_fmt",  HAS_ARG | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT | OPT_STRING, { .off = OFFSET(enc_stats_pre_fmt)  },
+            "format of the stats written with -stats_enc_pre" },
+        { "stats_enc_post_fmt", HAS_ARG | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT | OPT_STRING, { .off = OFFSET(enc_stats_post_fmt) },
+            "format of the stats written with -stats_enc_post" },
+        { "stats_mux_pre_fmt",  HAS_ARG | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT | OPT_STRING, { .off = OFFSET(mux_stats_fmt)      },
+            "format of the stats written with -stats_mux_pre" },
+
+        /* video options */
+        { "vframes",      OPT_VIDEO | HAS_ARG  | OPT_PERFILE | OPT_OUTPUT,           { .func_arg = opt_video_frames },
+            "set the number of video frames to output", "number" },
+        { "r",            OPT_VIDEO | HAS_ARG  | OPT_STRING | OPT_SPEC |
+                          OPT_INPUT | OPT_OUTPUT,                                    { .off = OFFSET(frame_rates) },
+            "set frame rate (Hz value, fraction or abbreviation)", "rate" },
+        { "fpsmax",       OPT_VIDEO | HAS_ARG  | OPT_STRING | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(max_frame_rates) },
+            "set max frame rate (Hz value, fraction or abbreviation)", "rate" },
+        { "s",            OPT_VIDEO | HAS_ARG | OPT_SUBTITLE | OPT_STRING | OPT_SPEC |
+                          OPT_INPUT | OPT_OUTPUT,                                    { .off = OFFSET(frame_sizes) },
+            "set frame size (WxH or abbreviation)", "size" },
+        { "aspect",       OPT_VIDEO | HAS_ARG  | OPT_STRING | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(frame_aspect_ratios) },
+            "set aspect ratio (4:3, 16:9 or 1.3333, 1.7777)", "aspect" },
+        { "pix_fmt",      OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_STRING | OPT_SPEC |
+                          OPT_INPUT | OPT_OUTPUT,                                    { .off = OFFSET(frame_pix_fmts) },
+            "set pixel format", "format" },
+        { "display_rotation", OPT_VIDEO | HAS_ARG | OPT_DOUBLE | OPT_SPEC |
+                              OPT_INPUT,                                             { .off = OFFSET(display_rotations) },
+            "set pure counter-clockwise rotation in degrees for stream(s)",
+            "angle" },
+        { "display_hflip", OPT_VIDEO | OPT_BOOL | OPT_SPEC | OPT_INPUT,              { .off = OFFSET(display_hflips) },
+            "set display horizontal flip for stream(s) "
+            "(overrides any display rotation if it is not set)"},
+        { "display_vflip", OPT_VIDEO | OPT_BOOL | OPT_SPEC | OPT_INPUT,              { .off = OFFSET(display_vflips) },
+            "set display vertical flip for stream(s) "
+            "(overrides any display rotation if it is not set)"},
+        { "vn",           OPT_VIDEO | OPT_BOOL  | OPT_OFFSET | OPT_INPUT | OPT_OUTPUT,{ .off = OFFSET(video_disable) },
+            "disable video" },
+        { "rc_override",  OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_STRING | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(rc_overrides) },
+            "rate control override for specific intervals", "override" },
+        { "vcodec",       OPT_VIDEO | HAS_ARG  | OPT_PERFILE | OPT_INPUT |
+                          OPT_OUTPUT,                                                { .func_arg = opt_video_codec },
+            "force video codec ('copy' to copy stream)", "codec" },
+        { "timecode",     OPT_VIDEO | HAS_ARG | OPT_PERFILE | OPT_OUTPUT,            { .func_arg = opt_timecode },
+            "set initial TimeCode value.", "hh:mm:ss[:;.]ff" },
+        { "pass",         OPT_VIDEO | HAS_ARG | OPT_SPEC | OPT_INT | OPT_OUTPUT,     { .off = OFFSET(pass) },
+            "select the pass number (1 to 3)", "n" },
+        { "passlogfile",  OPT_VIDEO | HAS_ARG | OPT_STRING | OPT_EXPERT | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(passlogfiles) },
+            "select two pass log file name prefix", "prefix" },
+    #if FFMPEG_OPT_PSNR
+        { "psnr",         OPT_VIDEO | OPT_BOOL | OPT_EXPERT,                         { &do_psnr },
+            "calculate PSNR of compressed frames (deprecated, use -flags +psnr)" },
+    #endif
+        { "vstats",       OPT_VIDEO | OPT_EXPERT ,                                   { .func_arg = opt_vstats },
+            "dump video coding statistics to file" },
+        { "vstats_file",  OPT_VIDEO | HAS_ARG | OPT_EXPERT ,                         { .func_arg = opt_vstats_file },
+            "dump video coding statistics to file", "file" },
+        { "vstats_version",  OPT_VIDEO | OPT_INT | HAS_ARG | OPT_EXPERT ,            { &vstats_version },
+            "Version of the vstats format to use."},
+        { "vf",           OPT_VIDEO | HAS_ARG  | OPT_PERFILE | OPT_OUTPUT,           { .func_arg = opt_video_filters },
+            "set video filters", "filter_graph" },
+        { "intra_matrix", OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_STRING | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(intra_matrices) },
+            "specify intra matrix coeffs", "matrix" },
+        { "inter_matrix", OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_STRING | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(inter_matrices) },
+            "specify inter matrix coeffs", "matrix" },
+        { "chroma_intra_matrix", OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_STRING | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(chroma_intra_matrices) },
+            "specify intra matrix coeffs", "matrix" },
+        { "top",          OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_INT| OPT_SPEC |
+                          OPT_INPUT | OPT_OUTPUT,                                    { .off = OFFSET(top_field_first) },
+            "top=1/bottom=0/auto=-1 field first", "" },
+        { "vtag",         OPT_VIDEO | HAS_ARG | OPT_EXPERT  | OPT_PERFILE |
+                          OPT_INPUT | OPT_OUTPUT,                                    { .func_arg = opt_old2new },
+            "force video tag/fourcc", "fourcc/tag" },
+        { "qphist",       OPT_VIDEO | OPT_BOOL | OPT_EXPERT ,                        { &qp_hist },
+            "show QP histogram" },
+        { "fps_mode",     OPT_VIDEO | HAS_ARG | OPT_STRING | OPT_EXPERT |
+                          OPT_SPEC | OPT_OUTPUT,                                     { .off = OFFSET(fps_mode) },
+            "set framerate mode for matching video streams; overrides vsync" },
+        { "force_fps",    OPT_VIDEO | OPT_BOOL | OPT_EXPERT  | OPT_SPEC |
+                          OPT_OUTPUT,                                                { .off = OFFSET(force_fps) },
+            "force the selected framerate, disable the best supported framerate selection" },
+        { "streamid",     OPT_VIDEO | HAS_ARG | OPT_EXPERT | OPT_PERFILE |
+                          OPT_OUTPUT,                                                { .func_arg = opt_streamid },
+            "set the value of an outfile streamid", "streamIndex:value" },
+        { "force_key_frames", OPT_VIDEO | OPT_STRING | HAS_ARG | OPT_EXPERT |
+                              OPT_SPEC | OPT_OUTPUT,                                 { .off = OFFSET(forced_key_frames) },
+            "force key frames at specified timestamps", "timestamps" },
+        { "b",            OPT_VIDEO | HAS_ARG | OPT_PERFILE | OPT_OUTPUT,            { .func_arg = opt_bitrate },
+            "video bitrate (please use -b:v)", "bitrate" },
+        { "hwaccel",          OPT_VIDEO | OPT_STRING | HAS_ARG | OPT_EXPERT |
+                              OPT_SPEC | OPT_INPUT,                                  { .off = OFFSET(hwaccels) },
+            "use HW accelerated decoding", "hwaccel name" },
+        { "hwaccel_device",   OPT_VIDEO | OPT_STRING | HAS_ARG | OPT_EXPERT |
+                              OPT_SPEC | OPT_INPUT,                                  { .off = OFFSET(hwaccel_devices) },
+            "select a device for HW acceleration", "devicename" },
+        { "hwaccel_output_format", OPT_VIDEO | OPT_STRING | HAS_ARG | OPT_EXPERT |
+                              OPT_SPEC | OPT_INPUT,                                  { .off = OFFSET(hwaccel_output_formats) },
+            "select output format used with HW accelerated decoding", "format" },
+        { "hwaccels",         OPT_EXIT,                                              { .func_arg = show_hwaccels },
+            "show available HW acceleration methods" },
+        { "autorotate",       HAS_ARG | OPT_BOOL | OPT_SPEC |
+                              OPT_EXPERT | OPT_INPUT,                                { .off = OFFSET(autorotate) },
+            "automatically insert correct rotate filters" },
+        { "autoscale",        HAS_ARG | OPT_BOOL | OPT_SPEC |
+                              OPT_EXPERT | OPT_OUTPUT,                               { .off = OFFSET(autoscale) },
+            "automatically insert a scale filter at the end of the filter graph" },
+        { "fix_sub_duration_heartbeat", OPT_VIDEO | OPT_BOOL | OPT_EXPERT |
+                                        OPT_SPEC | OPT_OUTPUT,                       { .off = OFFSET(fix_sub_duration_heartbeat) },
+            "set this video output stream to be a heartbeat stream for "
+            "fix_sub_duration, according to which subtitles should be split at "
+            "random access points" },
+
+        /* audio options */
+        { "aframes",        OPT_AUDIO | HAS_ARG  | OPT_PERFILE | OPT_OUTPUT,           { .func_arg = opt_audio_frames },
+            "set the number of audio frames to output", "number" },
+        { "aq",             OPT_AUDIO | HAS_ARG  | OPT_PERFILE | OPT_OUTPUT,           { .func_arg = opt_audio_qscale },
+            "set audio quality (codec-specific)", "quality", },
+        { "ar",             OPT_AUDIO | HAS_ARG  | OPT_INT | OPT_SPEC |
+                            OPT_INPUT | OPT_OUTPUT,                                    { .off = OFFSET(audio_sample_rate) },
+            "set audio sampling rate (in Hz)", "rate" },
+        { "ac",             OPT_AUDIO | HAS_ARG  | OPT_INT | OPT_SPEC |
+                            OPT_INPUT | OPT_OUTPUT,                                    { .off = OFFSET(audio_channels) },
+            "set number of audio channels", "channels" },
+        { "an",             OPT_AUDIO | OPT_BOOL | OPT_OFFSET | OPT_INPUT | OPT_OUTPUT,{ .off = OFFSET(audio_disable) },
+            "disable audio" },
+        { "acodec",         OPT_AUDIO | HAS_ARG  | OPT_PERFILE |
+                            OPT_INPUT | OPT_OUTPUT,                                    { .func_arg = opt_audio_codec },
+            "force audio codec ('copy' to copy stream)", "codec" },
+        { "ab",             OPT_AUDIO | HAS_ARG | OPT_PERFILE | OPT_OUTPUT,            { .func_arg = opt_bitrate },
+            "audio bitrate (please use -b:a)", "bitrate" },
+        { "atag",           OPT_AUDIO | HAS_ARG  | OPT_EXPERT | OPT_PERFILE |
+                            OPT_OUTPUT,                                                { .func_arg = opt_old2new },
+            "force audio tag/fourcc", "fourcc/tag" },
+        { "sample_fmt",     OPT_AUDIO | HAS_ARG  | OPT_EXPERT | OPT_SPEC |
+                            OPT_STRING | OPT_INPUT | OPT_OUTPUT,                       { .off = OFFSET(sample_fmts) },
+            "set sample format", "format" },
+        { "channel_layout", OPT_AUDIO | HAS_ARG  | OPT_EXPERT | OPT_SPEC |
+                            OPT_STRING | OPT_INPUT | OPT_OUTPUT,                       { .off = OFFSET(audio_ch_layouts) },
+            "set channel layout", "layout" },
+        { "ch_layout",      OPT_AUDIO | HAS_ARG  | OPT_EXPERT | OPT_SPEC |
+                            OPT_STRING | OPT_INPUT | OPT_OUTPUT,                       { .off = OFFSET(audio_ch_layouts) },
+            "set channel layout", "layout" },
+        { "af",             OPT_AUDIO | HAS_ARG  | OPT_PERFILE | OPT_OUTPUT,           { .func_arg = opt_audio_filters },
+            "set audio filters", "filter_graph" },
+        { "guess_layout_max", OPT_AUDIO | HAS_ARG | OPT_INT | OPT_SPEC | OPT_EXPERT | OPT_INPUT, { .off = OFFSET(guess_layout_max) },
+          "set the maximum number of channels to try to guess the channel layout" },
+
+        /* subtitle options */
+        { "sn",     OPT_SUBTITLE | OPT_BOOL | OPT_OFFSET | OPT_INPUT | OPT_OUTPUT, { .off = OFFSET(subtitle_disable) },
+            "disable subtitle" },
+        { "scodec", OPT_SUBTITLE | HAS_ARG  | OPT_PERFILE | OPT_INPUT | OPT_OUTPUT, { .func_arg = opt_subtitle_codec },
+            "force subtitle codec ('copy' to copy stream)", "codec" },
+        { "stag",   OPT_SUBTITLE | HAS_ARG  | OPT_EXPERT  | OPT_PERFILE | OPT_OUTPUT, { .func_arg = opt_old2new }
+            , "force subtitle tag/fourcc", "fourcc/tag" },
+        { "fix_sub_duration", OPT_BOOL | OPT_EXPERT | OPT_SUBTITLE | OPT_SPEC | OPT_INPUT, { .off = OFFSET(fix_sub_duration) },
+            "fix subtitles duration" },
+        { "canvas_size", OPT_SUBTITLE | HAS_ARG | OPT_STRING | OPT_SPEC | OPT_INPUT, { .off = OFFSET(canvas_sizes) },
+            "set canvas size (WxH or abbreviation)", "size" },
+
+        /* muxer options */
+        { "muxdelay",   OPT_FLOAT | HAS_ARG | OPT_EXPERT | OPT_OFFSET | OPT_OUTPUT, { .off = OFFSET(mux_max_delay) },
+            "set the maximum demux-decode delay", "seconds" },
+        { "muxpreload", OPT_FLOAT | HAS_ARG | OPT_EXPERT | OPT_OFFSET | OPT_OUTPUT, { .off = OFFSET(mux_preload) },
+            "set the initial demux-decode delay", "seconds" },
+        { "sdp_file", HAS_ARG | OPT_EXPERT | OPT_OUTPUT, { .func_arg = opt_sdp_file },
+            "specify a file in which to print sdp information", "file" },
+
+        { "time_base", HAS_ARG | OPT_STRING | OPT_EXPERT | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(time_bases) },
+            "set the desired time base hint for output stream (1:24, 1:48000 or 0.04166, 2.0833e-5)", "ratio" },
+        { "enc_time_base", HAS_ARG | OPT_STRING | OPT_EXPERT | OPT_SPEC | OPT_OUTPUT, { .off = OFFSET(enc_time_bases) },
+            "set the desired time base for the encoder (1:24, 1:48000 or 0.04166, 2.0833e-5). "
+            "two special values are defined - "
+            "0 = use frame rate (video) or sample rate (audio),"
+            "-1 = match source time base", "ratio" },
+
+        { "bsf", HAS_ARG | OPT_STRING | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT, { .off = OFFSET(bitstream_filters) },
+            "A comma-separated list of bitstream filters", "bitstream_filters" },
+        { "absf", HAS_ARG | OPT_AUDIO | OPT_EXPERT| OPT_PERFILE | OPT_OUTPUT, { .func_arg = opt_old2new },
+            "deprecated", "audio bitstream_filters" },
+        { "vbsf", OPT_VIDEO | HAS_ARG | OPT_EXPERT| OPT_PERFILE | OPT_OUTPUT, { .func_arg = opt_old2new },
+            "deprecated", "video bitstream_filters" },
+
+        { "apre", HAS_ARG | OPT_AUDIO | OPT_EXPERT| OPT_PERFILE | OPT_OUTPUT,    { .func_arg = opt_preset },
+            "set the audio options to the indicated preset", "preset" },
+        { "vpre", OPT_VIDEO | HAS_ARG | OPT_EXPERT| OPT_PERFILE | OPT_OUTPUT,    { .func_arg = opt_preset },
+            "set the video options to the indicated preset", "preset" },
+        { "spre", HAS_ARG | OPT_SUBTITLE | OPT_EXPERT| OPT_PERFILE | OPT_OUTPUT, { .func_arg = opt_preset },
+            "set the subtitle options to the indicated preset", "preset" },
+        { "fpre", HAS_ARG | OPT_EXPERT| OPT_PERFILE | OPT_OUTPUT,                { .func_arg = opt_preset },
+            "set options from indicated preset file", "filename" },
+
+        { "max_muxing_queue_size", HAS_ARG | OPT_INT | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT, { .off = OFFSET(max_muxing_queue_size) },
+            "maximum number of packets that can be buffered while waiting for all streams to initialize", "packets" },
+        { "muxing_queue_data_threshold", HAS_ARG | OPT_INT | OPT_SPEC | OPT_EXPERT | OPT_OUTPUT, { .off = OFFSET(muxing_queue_data_threshold) },
+            "set the threshold after which max_muxing_queue_size is taken into account", "bytes" },
+
+        /* data codec support */
+        { "dcodec", HAS_ARG | OPT_DATA | OPT_PERFILE | OPT_EXPERT | OPT_INPUT | OPT_OUTPUT, { .func_arg = opt_data_codec },
+            "force data codec ('copy' to copy stream)", "codec" },
+        { "dn", OPT_BOOL | OPT_VIDEO | OPT_OFFSET | OPT_INPUT | OPT_OUTPUT, { .off = OFFSET(data_disable) },
+            "disable data" },
+
+    #if CONFIG_VAAPI
+        { "vaapi_device", HAS_ARG | OPT_EXPERT, { .func_arg = opt_vaapi_device },
+            "set VAAPI hardware device (DRM path or X11 display name)", "device" },
+    #endif
+
+    #if CONFIG_QSV
+        { "qsv_device", HAS_ARG | OPT_EXPERT, { .func_arg = opt_qsv_device },
+            "set QSV hardware device (DirectX adapter index, DRM path or X11 display name)", "device"},
+    #endif
+
+        { "init_hw_device", HAS_ARG | OPT_EXPERT, { .func_arg = opt_init_hw_device },
+            "initialise hardware device", "args" },
+        { "filter_hw_device", HAS_ARG | OPT_EXPERT, { .func_arg = opt_filter_hw_device },
+            "set hardware device used when filtering", "device" },
+
+        { NULL, },
+    };
+
+    ffmpeg_options = options;
+
+    int ret;
+    BenchmarkTimeStamps ti;
+
+    int savedCode = setjmp(ex_buf__);
+    if (savedCode == 0) {
+
+        ffmpeg_var_cleanup();
+
+        init_dynload();
+
+        register_exit(ffmpeg_cleanup);
+
+        av_log_set_flags(AV_LOG_SKIP_REPEATED);
+        parse_loglevel(argc, argv, options);
+
+    #if CONFIG_AVDEVICE
+        avdevice_register_all();
+    #endif
+        avformat_network_init();
+
+        show_banner(argc, argv, options);
+
+        /* parse options and open all input/output files */
+        ret = ffmpeg_parse_options(argc, argv);
+        if (ret < 0)
+            exit_program(1);
+
+        if (nb_output_files <= 0 && nb_input_files == 0) {
+            show_usage();
+            av_log(NULL, AV_LOG_WARNING, "Use -h to get full help or, even better, run 'man %s'\n", program_name);
+            exit_program(1);
+        }
+
+        /* file converter / grab */
+        if (nb_output_files <= 0) {
+            av_log(NULL, AV_LOG_FATAL, "At least one output file must be specified\n");
+            exit_program(1);
+        }
+
+        current_time = ti = get_benchmark_time_stamps();
+        if (transcode() < 0)
+            exit_program(1);
+        if (do_benchmark) {
+            int64_t utime, stime, rtime;
+            current_time = get_benchmark_time_stamps();
+            utime = current_time.user_usec - ti.user_usec;
+            stime = current_time.sys_usec  - ti.sys_usec;
+            rtime = current_time.real_usec - ti.real_usec;
+            av_log(NULL, AV_LOG_INFO,
+                   "bench: utime=%0.3fs stime=%0.3fs rtime=%0.3fs\n",
+                   utime / 1000000.0, stime / 1000000.0, rtime / 1000000.0);
+        }
+        av_log(NULL, AV_LOG_DEBUG, "%"PRIu64" frames successfully decoded, %"PRIu64" decoding errors\n",
+               decode_error_stat[0], decode_error_stat[1]);
+        if ((decode_error_stat[0] + decode_error_stat[1]) * max_error_rate < decode_error_stat[1])
+            exit_program(69);
+
+        exit_program((received_nb_signals || cancelRequested(globalSessionId))? 255 : main_ffmpeg_return_code);
+
+    } else {
+        main_ffmpeg_return_code = (received_nb_signals || cancelRequested(globalSessionId)) ? 255 : longjmp_value;
+    }
+
+    return main_ffmpeg_return_code;
 }
