--- /tmp/ffmpeg-6.0-src/fftools/fopen_utf8.h	2026-04-06 17:05:14
+++ .fftools-7.1-work/kit60/fopen_utf8.h	2026-04-06 18:08:33
@@ -16,6 +16,14 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
+/*
+ * This file is the modified version of fopen_utf8.h file living in ffmpeg source code under the fftools folder. We
+ * manually update it each time we depend on a new ffmpeg version. Below you can see the list of changes applied
+ * by us to develop the ffmpeg-kit library.
+ *
+ * ffmpeg-kit changes by Taner Sener
+ */
+
 #ifndef FFTOOLS_FOPEN_UTF8_H
 #define FFTOOLS_FOPEN_UTF8_H
 
