--- /tmp/ffmpeg-6.0-src/fftools/objpool.c	2026-04-06 17:05:14
+++ .fftools-7.1-work/kit60/objpool.c	2026-04-06 18:08:33
@@ -1,5 +1,6 @@
 /*
  * This file is part of FFmpeg.
+ * Copyright (c) 2023 ARTHENICA LTD
  *
  * FFmpeg is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -16,6 +17,19 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
+/*
+ * This file is the modified version of objpool.c file living in ffmpeg source code under the fftools folder. We
+ * manually update it each time we depend on a new ffmpeg version. Below you can see the list of changes applied
+ * by us to develop ffmpeg-kit library.
+ *
+ * ffmpeg-kit changes by ARTHENICA LTD
+ *
+ * 07.2023
+ * --------------------------------------------------------
+ * - FFmpeg 6.0 changes migrated
+ * - fftools header names updated
+ */
+
 #include <stdint.h>
 
 #include "libavcodec/packet.h"
@@ -25,7 +39,7 @@
 #include "libavutil/frame.h"
 #include "libavutil/mem.h"
 
-#include "objpool.h"
+#include "fftools_objpool.h"
 
 struct ObjPool {
     void        *pool[32];
