--- /tmp/ffmpeg-6.0-src/fftools/opt_common.c	2026-04-06 17:05:14
+++ .fftools-7.1-work/kit60/opt_common.c	2026-04-06 18:08:33
@@ -1,5 +1,7 @@
 /*
  * Option handlers shared between the tools.
+ * Copyright (c) 2022 Taner Sener
+ * Copyright (c) 2023 ARTHENICA LTD
  *
  * This file is part of FFmpeg.
  *
@@ -18,12 +20,35 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
+/*
+ * This file is the modified version of opt_common.c file living in ffmpeg source code under the fftools folder. We
+ * manually update it each time we depend on a new ffmpeg version. Below you can see the list of changes applied
+ * by us to develop the ffmpeg-kit library.
+ *
+ * ffmpeg-kit changes by ARTHENICA LTD
+ *
+ * 07.2023
+ * --------------------------------------------------------
+ * - time field in report_callback updated as double
+ *
+ * ffmpeg-kit changes by Taner Sener
+ *
+ * 09.2022
+ * --------------------------------------------------------
+ * - printf replaced with av_log statements
+ * - libpostproc references dropped
+ * - fftools_ prefix added to fftools headers
+ * - added __thread modifier to report_file variables and warned_cfg
+ * - log_callback_report updated to forward logs via ffmpegkit_log_callback_function
+ * - extern program_name declared
+ */
+
 #include "config.h"
 
 #include <stdio.h>
 
-#include "cmdutils.h"
-#include "opt_common.h"
+#include "fftools_cmdutils.h"
+#include "fftools_opt_common.h"
 
 #include "libavutil/avassert.h"
 #include "libavutil/avstring.h"
@@ -60,27 +85,28 @@
 #include "libswresample/swresample.h"
 #include "libswresample/version.h"
 
-#include "libpostproc/postprocess.h"
-#include "libpostproc/version.h"
-
 enum show_muxdemuxers {
     SHOW_DEFAULT,
     SHOW_DEMUXERS,
     SHOW_MUXERS,
 };
 
-static FILE *report_file;
-static int report_file_level = AV_LOG_DEBUG;
+static __thread FILE *report_file = NULL;
+static __thread int report_file_level = AV_LOG_DEBUG;
 
+extern void ffmpegkit_log_callback_function(void *ptr, int level, const char* format, va_list vargs);
+extern void (*report_callback)(int, float, float, int64_t, double, double, double);
+extern __thread char *program_name;
+
 int show_license(void *optctx, const char *opt, const char *arg)
 {
 #if CONFIG_NONFREE
-    printf(
+    av_log(NULL, AV_LOG_ERROR,
     "This version of %s has nonfree parts compiled in.\n"
     "Therefore it is not legally redistributable.\n",
     program_name );
 #elif CONFIG_GPLV3
-    printf(
+    av_log(NULL, AV_LOG_ERROR,
     "%s is free software; you can redistribute it and/or modify\n"
     "it under the terms of the GNU General Public License as published by\n"
     "the Free Software Foundation; either version 3 of the License, or\n"
@@ -95,7 +121,7 @@
     "along with %s.  If not, see <http://www.gnu.org/licenses/>.\n",
     program_name, program_name, program_name );
 #elif CONFIG_GPL
-    printf(
+    av_log(NULL, AV_LOG_ERROR,
     "%s is free software; you can redistribute it and/or modify\n"
     "it under the terms of the GNU General Public License as published by\n"
     "the Free Software Foundation; either version 2 of the License, or\n"
@@ -111,7 +137,7 @@
     "Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA\n",
     program_name, program_name, program_name );
 #elif CONFIG_LGPLV3
-    printf(
+    av_log(NULL, AV_LOG_ERROR,
     "%s is free software; you can redistribute it and/or modify\n"
     "it under the terms of the GNU Lesser General Public License as published by\n"
     "the Free Software Foundation; either version 3 of the License, or\n"
@@ -126,7 +152,7 @@
     "along with %s.  If not, see <http://www.gnu.org/licenses/>.\n",
     program_name, program_name, program_name );
 #else
-    printf(
+    av_log(NULL, AV_LOG_ERROR,
     "%s is free software; you can redistribute it and/or\n"
     "modify it under the terms of the GNU Lesser General Public\n"
     "License as published by the Free Software Foundation; either\n"
@@ -146,7 +172,7 @@
     return 0;
 }
 
-static int warned_cfg = 0;
+__thread int warned_cfg = 0;
 
 #define INDENT        1
 #define SHOW_VERSION  2
@@ -191,7 +217,6 @@
     PRINT_LIB_INFO(avfilter,   AVFILTER,   flags, level);
     PRINT_LIB_INFO(swscale,    SWSCALE,    flags, level);
     PRINT_LIB_INFO(swresample, SWRESAMPLE, flags, level);
-    PRINT_LIB_INFO(postproc,   POSTPROC,   flags, level);
 }
 
 static void print_program_info(int flags, int level)
@@ -247,7 +272,6 @@
 
 int show_version(void *optctx, const char *opt, const char *arg)
 {
-    av_log_set_callback(log_callback_help);
     print_program_info (SHOW_COPYRIGHT, AV_LOG_INFO);
     print_all_libs_info(SHOW_VERSION, AV_LOG_INFO);
 
@@ -256,7 +280,6 @@
 
 int show_buildconf(void *optctx, const char *opt, const char *arg)
 {
-    av_log_set_callback(log_callback_help);
     print_buildconf      (INDENT|0, AV_LOG_INFO);
 
     return 0;
@@ -266,94 +289,91 @@
     if (codec->field) {                                                      \
         const type *p = codec->field;                                        \
                                                                              \
-        printf("    Supported " list_name ":");                              \
+        av_log(NULL, AV_LOG_ERROR, "    Supported " list_name ":");                              \
         while (*p != term) {                                                 \
             get_name(*p);                                                    \
-            printf(" %s", name);                                             \
+            av_log(NULL, AV_LOG_ERROR, " %s", name);                                             \
             p++;                                                             \
         }                                                                    \
-        printf("\n");                                                        \
+        av_log(NULL, AV_LOG_ERROR, "\n");                                                        \
     }                                                                        \
 
 static void print_codec(const AVCodec *c)
 {
     int encoder = av_codec_is_encoder(c);
 
-    printf("%s %s [%s]:\n", encoder ? "Encoder" : "Decoder", c->name,
+    av_log(NULL, AV_LOG_ERROR, "%s %s [%s]:\n", encoder ? "Encoder" : "Decoder", c->name,
            c->long_name ? c->long_name : "");
 
-    printf("    General capabilities: ");
+    av_log(NULL, AV_LOG_ERROR, "    General capabilities: ");
     if (c->capabilities & AV_CODEC_CAP_DRAW_HORIZ_BAND)
-        printf("horizband ");
+        av_log(NULL, AV_LOG_ERROR, "horizband ");
     if (c->capabilities & AV_CODEC_CAP_DR1)
-        printf("dr1 ");
+        av_log(NULL, AV_LOG_ERROR, "dr1 ");
     if (c->capabilities & AV_CODEC_CAP_DELAY)
-        printf("delay ");
+        av_log(NULL, AV_LOG_ERROR, "delay ");
     if (c->capabilities & AV_CODEC_CAP_SMALL_LAST_FRAME)
-        printf("small ");
+        av_log(NULL, AV_LOG_ERROR, "small ");
     if (c->capabilities & AV_CODEC_CAP_SUBFRAMES)
-        printf("subframes ");
+        av_log(NULL, AV_LOG_ERROR, "subframes ");
     if (c->capabilities & AV_CODEC_CAP_EXPERIMENTAL)
-        printf("exp ");
+        av_log(NULL, AV_LOG_ERROR, "exp ");
     if (c->capabilities & AV_CODEC_CAP_CHANNEL_CONF)
-        printf("chconf ");
+        av_log(NULL, AV_LOG_ERROR, "chconf ");
     if (c->capabilities & AV_CODEC_CAP_PARAM_CHANGE)
-        printf("paramchange ");
+        av_log(NULL, AV_LOG_ERROR, "paramchange ");
     if (c->capabilities & AV_CODEC_CAP_VARIABLE_FRAME_SIZE)
-        printf("variable ");
+        av_log(NULL, AV_LOG_ERROR, "variable ");
     if (c->capabilities & (AV_CODEC_CAP_FRAME_THREADS |
                            AV_CODEC_CAP_SLICE_THREADS |
                            AV_CODEC_CAP_OTHER_THREADS))
-        printf("threads ");
+        av_log(NULL, AV_LOG_ERROR, "threads ");
     if (c->capabilities & AV_CODEC_CAP_AVOID_PROBING)
-        printf("avoidprobe ");
+        av_log(NULL, AV_LOG_ERROR, "avoidprobe ");
     if (c->capabilities & AV_CODEC_CAP_HARDWARE)
-        printf("hardware ");
+        av_log(NULL, AV_LOG_ERROR, "hardware ");
     if (c->capabilities & AV_CODEC_CAP_HYBRID)
-        printf("hybrid ");
+        av_log(NULL, AV_LOG_ERROR, "hybrid ");
     if (!c->capabilities)
-        printf("none");
-    printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "none");
+    av_log(NULL, AV_LOG_ERROR, "\n");
 
     if (c->type == AVMEDIA_TYPE_VIDEO ||
         c->type == AVMEDIA_TYPE_AUDIO) {
-        printf("    Threading capabilities: ");
+        av_log(NULL, AV_LOG_ERROR, "    Threading capabilities: ");
         switch (c->capabilities & (AV_CODEC_CAP_FRAME_THREADS |
                                    AV_CODEC_CAP_SLICE_THREADS |
                                    AV_CODEC_CAP_OTHER_THREADS)) {
         case AV_CODEC_CAP_FRAME_THREADS |
-             AV_CODEC_CAP_SLICE_THREADS: printf("frame and slice"); break;
-        case AV_CODEC_CAP_FRAME_THREADS: printf("frame");           break;
-        case AV_CODEC_CAP_SLICE_THREADS: printf("slice");           break;
-        case AV_CODEC_CAP_OTHER_THREADS: printf("other");           break;
-        default:                         printf("none");            break;
+             AV_CODEC_CAP_SLICE_THREADS: av_log(NULL, AV_LOG_ERROR, "frame and slice"); break;
+        case AV_CODEC_CAP_FRAME_THREADS: av_log(NULL, AV_LOG_ERROR, "frame");           break;
+        case AV_CODEC_CAP_SLICE_THREADS: av_log(NULL, AV_LOG_ERROR, "slice");           break;
+        case AV_CODEC_CAP_OTHER_THREADS: av_log(NULL, AV_LOG_ERROR, "other");           break;
+        default:                         av_log(NULL, AV_LOG_ERROR, "none");            break;
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "\n");
     }
 
     if (avcodec_get_hw_config(c, 0)) {
-        printf("    Supported hardware devices: ");
+        av_log(NULL, AV_LOG_ERROR, "    Supported hardware devices: ");
         for (int i = 0;; i++) {
             const AVCodecHWConfig *config = avcodec_get_hw_config(c, i);
-            const char *name;
             if (!config)
                 break;
-            name = av_hwdevice_get_type_name(config->device_type);
-            if (name)
-                printf("%s ", name);
+            av_log(NULL, AV_LOG_ERROR, "%s ", av_hwdevice_get_type_name(config->device_type));
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "\n");
     }
 
     if (c->supported_framerates) {
         const AVRational *fps = c->supported_framerates;
 
-        printf("    Supported framerates:");
+        av_log(NULL, AV_LOG_ERROR, "    Supported framerates:");
         while (fps->num) {
-            printf(" %d/%d", fps->num, fps->den);
+            av_log(NULL, AV_LOG_ERROR, " %d/%d", fps->num, fps->den);
             fps++;
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "\n");
     }
     PRINT_CODEC_SUPPORTED(c, pix_fmts, enum AVPixelFormat, "pixel formats",
                           AV_PIX_FMT_NONE, GET_PIX_FMT_NAME);
@@ -365,14 +385,14 @@
     if (c->ch_layouts) {
         const AVChannelLayout *p = c->ch_layouts;
 
-        printf("    Supported channel layouts:");
+        av_log(NULL, AV_LOG_ERROR, "    Supported channel layouts:");
         while (p->nb_channels) {
             char name[128];
             av_channel_layout_describe(p, name, sizeof(name));
-            printf(" %s", name);
+            av_log(NULL, AV_LOG_ERROR, " %s", name);
             p++;
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "\n");
     }
 
     if (c->priv_class) {
@@ -439,10 +459,10 @@
         return;
     }
 
-    printf("Demuxer %s [%s]:\n", fmt->name, fmt->long_name);
+    av_log(NULL, AV_LOG_ERROR, "Demuxer %s [%s]:\n", fmt->name, fmt->long_name);
 
     if (fmt->extensions)
-        printf("    Common extensions: %s.\n", fmt->extensions);
+        av_log(NULL, AV_LOG_ERROR, "    Common extensions: %s.\n", fmt->extensions);
 
     if (fmt->priv_class)
         show_help_children(fmt->priv_class, AV_OPT_FLAG_DECODING_PARAM);
@@ -476,23 +496,23 @@
         return;
     }
 
-    printf("Muxer %s [%s]:\n", fmt->name, fmt->long_name);
+    av_log(NULL, AV_LOG_ERROR, "Muxer %s [%s]:\n", fmt->name, fmt->long_name);
 
     if (fmt->extensions)
-        printf("    Common extensions: %s.\n", fmt->extensions);
+        av_log(NULL, AV_LOG_ERROR, "    Common extensions: %s.\n", fmt->extensions);
     if (fmt->mime_type)
-        printf("    Mime type: %s.\n", fmt->mime_type);
+        av_log(NULL, AV_LOG_ERROR, "    Mime type: %s.\n", fmt->mime_type);
     if (fmt->video_codec != AV_CODEC_ID_NONE &&
         (desc = avcodec_descriptor_get(fmt->video_codec))) {
-        printf("    Default video codec: %s.\n", desc->name);
+        av_log(NULL, AV_LOG_ERROR, "    Default video codec: %s.\n", desc->name);
     }
     if (fmt->audio_codec != AV_CODEC_ID_NONE &&
         (desc = avcodec_descriptor_get(fmt->audio_codec))) {
-        printf("    Default audio codec: %s.\n", desc->name);
+        av_log(NULL, AV_LOG_ERROR, "    Default audio codec: %s.\n", desc->name);
     }
     if (fmt->subtitle_codec != AV_CODEC_ID_NONE &&
         (desc = avcodec_descriptor_get(fmt->subtitle_codec))) {
-        printf("    Default subtitle codec: %s.\n", desc->name);
+        av_log(NULL, AV_LOG_ERROR, "    Default subtitle codec: %s.\n", desc->name);
     }
 
     if (fmt->priv_class)
@@ -514,40 +534,40 @@
         return;
     }
 
-    printf("Filter %s\n", f->name);
+    av_log(NULL, AV_LOG_ERROR, "Filter %s\n", f->name);
     if (f->description)
-        printf("  %s\n", f->description);
+        av_log(NULL, AV_LOG_ERROR, "  %s\n", f->description);
 
     if (f->flags & AVFILTER_FLAG_SLICE_THREADS)
-        printf("    slice threading supported\n");
+        av_log(NULL, AV_LOG_ERROR, "    slice threading supported\n");
 
-    printf("    Inputs:\n");
+    av_log(NULL, AV_LOG_ERROR, "    Inputs:\n");
     count = avfilter_filter_pad_count(f, 0);
     for (i = 0; i < count; i++) {
-        printf("       #%d: %s (%s)\n", i, avfilter_pad_get_name(f->inputs, i),
+        av_log(NULL, AV_LOG_ERROR, "       #%d: %s (%s)\n", i, avfilter_pad_get_name(f->inputs, i),
                av_get_media_type_string(avfilter_pad_get_type(f->inputs, i)));
     }
     if (f->flags & AVFILTER_FLAG_DYNAMIC_INPUTS)
-        printf("        dynamic (depending on the options)\n");
+        av_log(NULL, AV_LOG_ERROR, "        dynamic (depending on the options)\n");
     else if (!count)
-        printf("        none (source filter)\n");
+        av_log(NULL, AV_LOG_ERROR, "        none (source filter)\n");
 
-    printf("    Outputs:\n");
+    av_log(NULL, AV_LOG_ERROR, "    Outputs:\n");
     count = avfilter_filter_pad_count(f, 1);
     for (i = 0; i < count; i++) {
-        printf("       #%d: %s (%s)\n", i, avfilter_pad_get_name(f->outputs, i),
+        av_log(NULL, AV_LOG_ERROR, "       #%d: %s (%s)\n", i, avfilter_pad_get_name(f->outputs, i),
                av_get_media_type_string(avfilter_pad_get_type(f->outputs, i)));
     }
     if (f->flags & AVFILTER_FLAG_DYNAMIC_OUTPUTS)
-        printf("        dynamic (depending on the options)\n");
+        av_log(NULL, AV_LOG_ERROR, "        dynamic (depending on the options)\n");
     else if (!count)
-        printf("        none (sink filter)\n");
+        av_log(NULL, AV_LOG_ERROR, "        none (sink filter)\n");
 
     if (f->priv_class)
         show_help_children(f->priv_class, AV_OPT_FLAG_VIDEO_PARAM | AV_OPT_FLAG_FILTERING_PARAM |
                                           AV_OPT_FLAG_AUDIO_PARAM);
     if (f->flags & AVFILTER_FLAG_SUPPORT_TIMELINE)
-        printf("This filter has support for timeline through the 'enable' option.\n");
+        av_log(NULL, AV_LOG_ERROR, "This filter has support for timeline through the 'enable' option.\n");
 #else
     av_log(NULL, AV_LOG_ERROR, "Build without libavfilter; "
            "can not to satisfy request\n");
@@ -567,7 +587,7 @@
         return;
     }
 
-    printf("Bit stream filter %s\n", bsf->name);
+    av_log(NULL, AV_LOG_ERROR, "Bit stream filter %s\n", bsf->name);
     PRINT_CODEC_SUPPORTED(bsf, codec_ids, enum AVCodecID, "codecs",
                           AV_CODEC_ID_NONE, GET_CODEC_NAME);
     if (bsf->priv_class)
@@ -577,7 +597,6 @@
 int show_help(void *optctx, const char *opt, const char *arg)
 {
     char *topic, *par;
-    av_log_set_callback(log_callback_help);
 
     topic = av_strdup(arg ? arg : "");
     if (!topic)
@@ -587,7 +606,11 @@
         *par++ = 0;
 
     if (!*topic) {
-        show_help_default(topic, par);
+        if (program_name && !strcmp(program_name, "ffmpeg")) {
+            show_help_default_ffmpeg(topic, par);
+        } else {
+            show_help_default_ffprobe(topic, par);
+        }
     } else if (!strcmp(topic, "decoder")) {
         show_help_codec(par, 0);
     } else if (!strcmp(topic, "encoder")) {
@@ -605,7 +628,11 @@
     } else if (!strcmp(topic, "bsf")) {
         show_help_bsf(par);
     } else {
-        show_help_default(topic, par);
+        if (program_name && !strcmp(program_name, "ffmpeg")) {
+            show_help_default_ffmpeg(topic, par);
+        } else {
+            show_help_default_ffprobe(topic, par);
+        }
     }
 
     av_freep(&topic);
@@ -617,12 +644,12 @@
     void *iter = NULL;
     const AVCodec *codec;
 
-    printf(" (%s: ", encoder ? "encoders" : "decoders");
+    av_log(NULL, AV_LOG_ERROR, " (%s: ", encoder ? "encoders" : "decoders");
 
     while ((codec = next_codec_for_id(id, &iter, encoder)))
-        printf("%s ", codec->name);
+        av_log(NULL, AV_LOG_ERROR, "%s ", codec->name);
 
-    printf(")");
+    av_log(NULL, AV_LOG_ERROR, ")");
 }
 
 static int compare_codec_desc(const void *a, const void *b)
@@ -642,8 +669,10 @@
 
     while ((desc = avcodec_descriptor_next(desc)))
         nb_codecs++;
-    if (!(codecs = av_calloc(nb_codecs, sizeof(*codecs))))
-        report_and_exit(AVERROR(ENOMEM));
+    if (!(codecs = av_calloc(nb_codecs, sizeof(*codecs)))) {
+        av_log(NULL, AV_LOG_ERROR, "Out of memory\n");
+        exit_program(1);
+    }
     desc = NULL;
     while ((desc = avcodec_descriptor_next(desc)))
         codecs[i++] = desc;
@@ -670,7 +699,7 @@
     const AVCodecDescriptor **codecs;
     unsigned i, nb_codecs = get_codecs_sorted(&codecs);
 
-    printf("Codecs:\n"
+    av_log(NULL, AV_LOG_ERROR, "Codecs:\n"
            " D..... = Decoding supported\n"
            " .E.... = Encoding supported\n"
            " ..V... = Video codec\n"
@@ -690,16 +719,16 @@
         if (strstr(desc->name, "_deprecated"))
             continue;
 
-        printf(" ");
-        printf(avcodec_find_decoder(desc->id) ? "D" : ".");
-        printf(avcodec_find_encoder(desc->id) ? "E" : ".");
+        av_log(NULL, AV_LOG_ERROR, " ");
+        av_log(NULL, AV_LOG_ERROR, avcodec_find_decoder(desc->id) ? "D" : ".");
+        av_log(NULL, AV_LOG_ERROR, avcodec_find_encoder(desc->id) ? "E" : ".");
 
-        printf("%c", get_media_type_char(desc->type));
-        printf((desc->props & AV_CODEC_PROP_INTRA_ONLY) ? "I" : ".");
-        printf((desc->props & AV_CODEC_PROP_LOSSY)      ? "L" : ".");
-        printf((desc->props & AV_CODEC_PROP_LOSSLESS)   ? "S" : ".");
+        av_log(NULL, AV_LOG_ERROR, "%c", get_media_type_char(desc->type));
+        av_log(NULL, AV_LOG_ERROR, (desc->props & AV_CODEC_PROP_INTRA_ONLY) ? "I" : ".");
+        av_log(NULL, AV_LOG_ERROR, (desc->props & AV_CODEC_PROP_LOSSY)      ? "L" : ".");
+        av_log(NULL, AV_LOG_ERROR, (desc->props & AV_CODEC_PROP_LOSSLESS)   ? "S" : ".");
 
-        printf(" %-20s %s", desc->name, desc->long_name ? desc->long_name : "");
+        av_log(NULL, AV_LOG_ERROR, " %-20s %s", desc->name, desc->long_name ? desc->long_name : "");
 
         /* print decoders/encoders when there's more than one or their
          * names are different from codec name */
@@ -717,7 +746,7 @@
             }
         }
 
-        printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "\n");
     }
     av_free(codecs);
     return 0;
@@ -728,7 +757,7 @@
     const AVCodecDescriptor **codecs;
     unsigned i, nb_codecs = get_codecs_sorted(&codecs);
 
-    printf("%s:\n"
+    av_log(NULL, AV_LOG_ERROR, "%s:\n"
            " V..... = Video\n"
            " A..... = Audio\n"
            " S..... = Subtitle\n"
@@ -745,18 +774,18 @@
         void *iter = NULL;
 
         while ((codec = next_codec_for_id(desc->id, &iter, encoder))) {
-            printf(" %c", get_media_type_char(desc->type));
-            printf((codec->capabilities & AV_CODEC_CAP_FRAME_THREADS) ? "F" : ".");
-            printf((codec->capabilities & AV_CODEC_CAP_SLICE_THREADS) ? "S" : ".");
-            printf((codec->capabilities & AV_CODEC_CAP_EXPERIMENTAL)  ? "X" : ".");
-            printf((codec->capabilities & AV_CODEC_CAP_DRAW_HORIZ_BAND)?"B" : ".");
-            printf((codec->capabilities & AV_CODEC_CAP_DR1)           ? "D" : ".");
+            av_log(NULL, AV_LOG_ERROR, " %c", get_media_type_char(desc->type));
+            av_log(NULL, AV_LOG_ERROR, (codec->capabilities & AV_CODEC_CAP_FRAME_THREADS) ? "F" : ".");
+            av_log(NULL, AV_LOG_ERROR, (codec->capabilities & AV_CODEC_CAP_SLICE_THREADS) ? "S" : ".");
+            av_log(NULL, AV_LOG_ERROR, (codec->capabilities & AV_CODEC_CAP_EXPERIMENTAL)  ? "X" : ".");
+            av_log(NULL, AV_LOG_ERROR, (codec->capabilities & AV_CODEC_CAP_DRAW_HORIZ_BAND)?"B" : ".");
+            av_log(NULL, AV_LOG_ERROR, (codec->capabilities & AV_CODEC_CAP_DR1)           ? "D" : ".");
 
-            printf(" %-20s %s", codec->name, codec->long_name ? codec->long_name : "");
+            av_log(NULL, AV_LOG_ERROR, " %-20s %s", codec->name, codec->long_name ? codec->long_name : "");
             if (strcmp(codec->name, desc->name))
-                printf(" (codec %s)", desc->name);
+                av_log(NULL, AV_LOG_ERROR, " (codec %s)", desc->name);
 
-            printf("\n");
+            av_log(NULL, AV_LOG_ERROR, "\n");
         }
     }
     av_free(codecs);
@@ -779,10 +808,10 @@
     const AVBitStreamFilter *bsf = NULL;
     void *opaque = NULL;
 
-    printf("Bitstream filters:\n");
+    av_log(NULL, AV_LOG_ERROR, "Bitstream filters:\n");
     while ((bsf = av_bsf_iterate(&opaque)))
-        printf("%s\n", bsf->name);
-    printf("\n");
+        av_log(NULL, AV_LOG_ERROR, "%s\n", bsf->name);
+    av_log(NULL, AV_LOG_ERROR, "\n");
     return 0;
 }
 
@@ -795,7 +824,7 @@
     int i, j;
     const AVFilterPad *pad;
 
-    printf("Filters:\n"
+    av_log(NULL, AV_LOG_ERROR, "Filters:\n"
            "  T.. = Timeline support\n"
            "  .S. = Slice threading\n"
            "  ..C = Command support\n"
@@ -823,14 +852,14 @@
                                   ( i && (filter->flags & AVFILTER_FLAG_DYNAMIC_OUTPUTS))) ? 'N' : '|';
         }
         *descr_cur = 0;
-        printf(" %c%c%c %-17s %-10s %s\n",
+        av_log(NULL, AV_LOG_ERROR, " %c%c%c %-17s %-10s %s\n",
                filter->flags & AVFILTER_FLAG_SUPPORT_TIMELINE ? 'T' : '.',
                filter->flags & AVFILTER_FLAG_SLICE_THREADS    ? 'S' : '.',
                filter->process_command                        ? 'C' : '.',
                filter->name, descr, filter->description);
     }
 #else
-    printf("No filters available: libavfilter disabled\n");
+    av_log(NULL, AV_LOG_ERROR, "No filters available: libavfilter disabled\n");
 #endif
     return 0;
 }
@@ -851,7 +880,7 @@
     const char *last_name;
     int is_dev;
 
-    printf("%s\n"
+    av_log(NULL, AV_LOG_ERROR, "%s\n"
            " D. = Demuxing supported\n"
            " .E = Muxing supported\n"
            " --\n", device_only ? "Devices:" : "File formats:");
@@ -896,7 +925,7 @@
             break;
         last_name = name;
 
-        printf(" %c%c %-15s %s\n",
+        av_log(NULL, AV_LOG_ERROR, " %c%c %-15s %s\n",
                decode ? 'D' : ' ',
                encode ? 'E' : ' ',
                name,
@@ -930,13 +959,13 @@
     void *opaque = NULL;
     const char *name;
 
-    printf("Supported file protocols:\n"
+    av_log(NULL, AV_LOG_ERROR, "Supported file protocols:\n"
            "Input:\n");
     while ((name = avio_enum_protocols(&opaque, 0)))
-        printf("  %s\n", name);
-    printf("Output:\n");
+        av_log(NULL, AV_LOG_ERROR, "  %s\n", name);
+    av_log(NULL, AV_LOG_ERROR, "Output:\n");
     while ((name = avio_enum_protocols(&opaque, 1)))
-        printf("  %s\n", name);
+        av_log(NULL, AV_LOG_ERROR, "  %s\n", name);
     return 0;
 }
 
@@ -946,10 +975,10 @@
     const uint8_t *rgb;
     int i;
 
-    printf("%-32s #RRGGBB\n", "name");
+    av_log(NULL, AV_LOG_ERROR, "%-32s #RRGGBB\n", "name");
 
-    for (i = 0; name = av_get_known_color_name(i, &rgb); i++)
-        printf("%-32s #%02x%02x%02x\n", name, rgb[0], rgb[1], rgb[2]);
+    for (i = 0; (name = av_get_known_color_name(i, &rgb)); i++)
+        av_log(NULL, AV_LOG_ERROR, "%-32s #%02x%02x%02x\n", name, rgb[0], rgb[1], rgb[2]);
 
     return 0;
 }
@@ -958,7 +987,7 @@
 {
     const AVPixFmtDescriptor *pix_desc = NULL;
 
-    printf("Pixel formats:\n"
+    av_log(NULL, AV_LOG_ERROR, "Pixel formats:\n"
            "I.... = Supported Input  format for conversion\n"
            ".O... = Supported Output format for conversion\n"
            "..H.. = Hardware accelerated format\n"
@@ -974,7 +1003,7 @@
 
     while ((pix_desc = av_pix_fmt_desc_next(pix_desc))) {
         enum AVPixelFormat av_unused pix_fmt = av_pix_fmt_desc_get_id(pix_desc);
-        printf("%c%c%c%c%c %-16s       %d            %3d      %d",
+        av_log(NULL, AV_LOG_ERROR, "%c%c%c%c%c %-16s       %d            %3d      %d",
                sws_isSupportedInput (pix_fmt)              ? 'I' : '.',
                sws_isSupportedOutput(pix_fmt)              ? 'O' : '.',
                pix_desc->flags & AV_PIX_FMT_FLAG_HWACCEL   ? 'H' : '.',
@@ -986,8 +1015,8 @@
                pix_desc->comp[0].depth);
 
         for (unsigned i = 1; i < pix_desc->nb_components; i++)
-            printf("-%d", pix_desc->comp[i].depth);
-        printf("\n");
+            av_log(NULL, AV_LOG_ERROR, "-%d", pix_desc->comp[i].depth);
+        av_log(NULL, AV_LOG_ERROR, "\n");
     }
     return 0;
 }
@@ -999,28 +1028,28 @@
     char buf[128], buf2[128];
     int i = 0;
 
-    printf("Individual channels:\n"
+    av_log(NULL, AV_LOG_ERROR, "Individual channels:\n"
            "NAME           DESCRIPTION\n");
     for (i = 0; i < 63; i++) {
         av_channel_name(buf, sizeof(buf), i);
         if (strstr(buf, "USR"))
             continue;
         av_channel_description(buf2, sizeof(buf2), i);
-        printf("%-14s %s\n", buf, buf2);
+        av_log(NULL, AV_LOG_ERROR, "%-14s %s\n", buf, buf2);
     }
-    printf("\nStandard channel layouts:\n"
+    av_log(NULL, AV_LOG_ERROR, "\nStandard channel layouts:\n"
            "NAME           DECOMPOSITION\n");
-    while (ch_layout = av_channel_layout_standard(&iter)) {
+    while ((ch_layout = av_channel_layout_standard(&iter))) {
             av_channel_layout_describe(ch_layout, buf, sizeof(buf));
-            printf("%-14s ", buf);
+            av_log(NULL, AV_LOG_ERROR, "%-14s ", buf);
             for (i = 0; i < 63; i++) {
                 int idx = av_channel_layout_index_from_channel(ch_layout, i);
                 if (idx >= 0) {
                     av_channel_name(buf2, sizeof(buf2), i);
-                    printf("%s%s", idx ? "+" : "", buf2);
+                    av_log(NULL, AV_LOG_ERROR, "%s%s", idx ? "+" : "", buf2);
                 }
             }
-            printf("\n");
+            av_log(NULL, AV_LOG_ERROR, "\n");
     }
     return 0;
 }
@@ -1030,7 +1059,7 @@
     int i;
     char fmt_str[128];
     for (i = -1; i < AV_SAMPLE_FMT_NB; i++)
-        printf("%s\n", av_get_sample_fmt_string(fmt_str, sizeof(fmt_str), i));
+        av_log(NULL, AV_LOG_ERROR, "%s\n", av_get_sample_fmt_string(fmt_str, sizeof(fmt_str), i));
     return 0;
 }
 
@@ -1039,7 +1068,7 @@
     for (int i = 0; i < 32; i++) {
         const char *str = av_disposition_to_string(1U << i);
         if (str)
-            printf("%s\n", str);
+            av_log(NULL, AV_LOG_ERROR, "%s\n", str);
     }
     return 0;
 }
@@ -1117,10 +1146,14 @@
     static int print_prefix = 1;
 
     va_copy(vl2, vl);
-    av_log_default_callback(ptr, level, fmt, vl);
+    if (report_callback == NULL) {
+        av_log_default_callback(ptr, level, fmt, vl);
+    } else {
+        ffmpegkit_log_callback_function(ptr, level, fmt, vl);
+    }
     av_log_format_line(ptr, level, fmt, vl2, line, sizeof(line), &print_prefix);
     va_end(vl2);
-    if (report_file_level >= level) {
+    if (report_file && report_file_level >= level) {
         fputs(line, report_file);
         fflush(report_file);
     }
@@ -1310,19 +1343,19 @@
     // print devices
     for (int i = 0; i < device_list->nb_devices; i++) {
         const AVDeviceInfo *device = device_list->devices[i];
-        printf("%c %s [%s] (", device_list->default_device == i ? '*' : ' ',
+        av_log(NULL, AV_LOG_ERROR, "%c %s [%s] (", device_list->default_device == i ? '*' : ' ',
             device->device_name, device->device_description);
         if (device->nb_media_types > 0) {
             for (int j = 0; j < device->nb_media_types; ++j) {
                 const char* media_type = av_get_media_type_string(device->media_types[j]);
                 if (j > 0)
-                    printf(", ");
-                printf("%s", media_type ? media_type : "unknown");
+                    av_log(NULL, AV_LOG_ERROR, ", ");
+                av_log(NULL, AV_LOG_ERROR, "%s", media_type ? media_type : "unknown");
             }
         } else {
-            printf("none");
+            av_log(NULL, AV_LOG_ERROR, "none");
         }
-        printf(")\n");
+        av_log(NULL, AV_LOG_ERROR, ")\n");
     }
 }
 
@@ -1334,9 +1367,9 @@
     if (!fmt || !fmt->priv_class  || !AV_IS_INPUT_DEVICE(fmt->priv_class->category))
         return AVERROR(EINVAL);
 
-    printf("Auto-detected sources for %s:\n", fmt->name);
+    av_log(NULL, AV_LOG_ERROR, "Auto-detected sources for %s:\n", fmt->name);
     if ((ret = avdevice_list_input_sources(fmt, NULL, opts, &device_list)) < 0) {
-        printf("Cannot list sources: %s\n", av_err2str(ret));
+        av_log(NULL, AV_LOG_ERROR, "Cannot list sources: %s\n", av_err2str(ret));
         goto fail;
     }
 
@@ -1355,9 +1388,9 @@
     if (!fmt || !fmt->priv_class  || !AV_IS_OUTPUT_DEVICE(fmt->priv_class->category))
         return AVERROR(EINVAL);
 
-    printf("Auto-detected sinks for %s:\n", fmt->name);
+    av_log(NULL, AV_LOG_ERROR, "Auto-detected sinks for %s:\n", fmt->name);
     if ((ret = avdevice_list_output_sinks(fmt, NULL, opts, &device_list)) < 0) {
-        printf("Cannot list sinks: %s\n", av_err2str(ret));
+        av_log(NULL, AV_LOG_ERROR, "Cannot list sinks: %s\n", av_err2str(ret));
         goto fail;
     }
 
@@ -1385,7 +1418,7 @@
             }
         }
     } else
-        printf("\nDevice name is not provided.\n"
+        av_log(NULL, AV_LOG_ERROR, "\nDevice name is not provided.\n"
                 "You can pass devicename[,opt1=val1[,opt2=val2...]] as an argument.\n\n");
     return 0;
 }
